{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0fd30be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"../../00_project_config.ipynb\"\n",
    "%run \"../00_path_config.ipynb\"\n",
    "\n",
    "import pathlib as pl\n",
    "import pandas as pd\n",
    "import re\n",
    "import itertools as itt\n",
    "\n",
    "prep_table = PROJECT_BASE.joinpath(\n",
    "    \"annotations\", \"data_freezes\",\n",
    "    \"out-110.chrY_sheet1_data_sources.prep.tsv\"\n",
    ")\n",
    "\n",
    "out_table = PROJECT_BASE.joinpath(\n",
    "    \"annotations\", \"data_freezes\",\n",
    "    \"out-120.hgsvc3_data_sources.prep.tsv\"\n",
    ")\n",
    "\n",
    "df = pd.read_csv(prep_table, sep=\"\\t\", header=0, comment=\"#\")\n",
    "\n",
    "unk_samples = ~df[\"sample\"].isin(HGSVC_SAMPLES[\"sample\"])\n",
    "unk_samples = sorted(df.loc[unk_samples, \"sample\"].values)\n",
    "\n",
    "df = df.loc[df[\"sample\"].isin(HGSVC_SAMPLES[\"sample\"]), :].copy()\n",
    "df.drop([\"population\", \"super_population\"], axis=1, inplace=True)\n",
    "\n",
    "full_table = HGSVC_SAMPLES.merge(df, left_on=\"sample\", right_on=\"sample\", how=\"outer\")\n",
    "full_table.rename(\n",
    "    {\n",
    "        \"order_num\": \"hgsvc_sample_number\",\n",
    "        \"batch_num\": \"verkko_assembly_batch\"\n",
    "    }, axis=1, inplace=True\n",
    ")\n",
    "full_table[\"member\"] = full_table[\"member\"].replace({\"-\": \"single\"})\n",
    "full_table[\"family\"] = full_table[\"family\"].replace({\"-\": \"single\"})\n",
    "full_table.loc[full_table[\"sample\"] == full_table[\"family\"], \"family\"] = \"single\"\n",
    "full_table = full_table.loc[full_table[\"verkko_assembly_batch\"] > 0, :].copy()\n",
    "\n",
    "def simplify_remote(remote_location):\n",
    "    \n",
    "    if pd.isnull(remote_location):\n",
    "        return \"unknown\"\n",
    "    loc = remote_location.strip()\n",
    "    if loc.startswith(\"rebasecalled\"):\n",
    "        loc = loc.split(\"-\")[1].strip()\n",
    "    loc = loc.strip(\"https:/\")\n",
    "    if \"PRJ\" in loc:\n",
    "        return \"unknown\"\n",
    "    return loc\n",
    "\n",
    "\n",
    "def simplify_accession(accession_num):\n",
    "    \n",
    "    if pd.isnull(accession_num):\n",
    "        return \"unknown\"\n",
    "    acc = re.match(\"PRJ[NAEB]+[0-9]+\", accession_num)\n",
    "    if acc is None:\n",
    "        if \"T2T\" in accession_num:\n",
    "            return \"unknown\"\n",
    "        raise ValueError(accession_num)\n",
    "    s, e = acc.span()\n",
    "    acc = accession_num[s:e]\n",
    "    return acc\n",
    "\n",
    "\n",
    "prefixes = {\n",
    "    \"hgsvc2\": \"IGSR:HGSVC2:\",\n",
    "    \"hgsvc3\": \"IGSR:HGSVC3:\",\n",
    "    \"awshprc\": \"AWS:HPRC:\"\n",
    "}\n",
    "\n",
    "prefix_paths = {\n",
    "    \"IGSR:HGSVC2:\": \"ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2\",\n",
    "    \"IGSR:HGSVC3:\": \"ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC3\",\n",
    "    \"AWS:HPRC:\": \"s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=\"\n",
    "}\n",
    "\n",
    "\n",
    "def prefix_remote(remote_url):\n",
    "    \n",
    "    if remote_url == \"unknown\":\n",
    "        prefix = \"\"\n",
    "        path = \"unknown\"\n",
    "    elif \"data_collections/HGSVC3\" in remote_url:\n",
    "        prefix = prefixes[\"hgsvc3\"]\n",
    "        path = remote_url.split(\"HGSVC3\")[-1]\n",
    "    elif \"data_collections/HGSVC2\" in remote_url:\n",
    "        prefix = prefixes[\"hgsvc2\"]\n",
    "        path = remote_url.split(\"HGSVC2\")[-1]\n",
    "    elif \"amazonaws.com/human-pangenomics\" in remote_url:\n",
    "        prefix = prefixes[\"awshprc\"]\n",
    "        path = remote_url.split(\"prefix=\")[-1]\n",
    "    elif \"github.com\" in remote_url:\n",
    "        prefix = \"\"\n",
    "        path = \"unknown\"\n",
    "    elif \"NA\" in remote_url:\n",
    "        prefix = \"\"\n",
    "        path = \"unknown\"\n",
    "    else:\n",
    "        raise ValueError(remote_url)\n",
    "    return f\"{prefix}{path}\"   \n",
    "\n",
    "\n",
    "def prefix_split_remotes_accession(data_row, datatype):\n",
    "    \n",
    "    assert datatype in [\"hifi\", \"ont\", \"strandseq\"]\n",
    "    sample = data_row.name\n",
    "    remote_fields = getattr(data_row, f\"remote_{datatype}\")\n",
    "    remote_fields = remote_fields.strip().split(\",\")    \n",
    "    remote_fields = [prefix_remote(remote.strip().strip(\"/\")) for remote in remote_fields]\n",
    "    \n",
    "    acc_fields = getattr(data_row, f\"accession_{datatype}\")\n",
    "    acc_fields = acc_fields.strip().split(\",\")\n",
    "    acc_fields = [acc.strip() for acc in acc_fields]\n",
    "    \n",
    "    fill_value = acc_fields[-1] if len(acc_fields) < len(remote_fields) else remote_fields[-1]\n",
    "    \n",
    "    fields = []\n",
    "    for remote, acc in itt.zip_longest(remote_fields, acc_fields, fillvalue=fill_value):\n",
    "        subfolder = remote.rsplit(\"/\")[-1]\n",
    "        if subfolder == \"fastq\" or subfolder == \"bam\":\n",
    "            subfolder = remote.rsplit(\"/\")[-2]\n",
    "        if subfolder[2:] in sample:\n",
    "            subfolder = remote.rsplit(\"/\")[-2]\n",
    "        fields.append((sample, datatype, remote, subfolder, acc))\n",
    "    return tuple(fields)\n",
    "\n",
    "\n",
    "full_table[\"remote_hifi\"] = full_table[\"remote_hifi\"].apply(simplify_remote)\n",
    "full_table[\"remote_ont\"] = full_table[\"remote_ont\"].apply(simplify_remote)\n",
    "full_table[\"remote_strandseq\"] = full_table[\"remote_strandseq\"].apply(simplify_remote)\n",
    "\n",
    "full_table[\"accession_hifi\"] = full_table[\"accession_hifi\"].apply(simplify_accession)\n",
    "full_table[\"accession_ont\"] = full_table[\"accession_ont\"].apply(simplify_accession)\n",
    "full_table[\"accession_strandseq\"] = full_table[\"accession_strandseq\"].apply(simplify_accession)\n",
    "\n",
    "full_table[\"SIN\"] = full_table[\"sample\"].apply(lambda x: f\"SIN:{x[2:]}\")\n",
    "\n",
    "full_table.set_index(\"SIN\", inplace=True)\n",
    "\n",
    "hifi_fields = list(\n",
    "    itt.chain(\n",
    "        *full_table.apply(prefix_split_remotes_accession, axis=1, args=(\"hifi\",)).values\n",
    "    ))\n",
    "ont_fields = list(\n",
    "    itt.chain(\n",
    "        *full_table.apply(prefix_split_remotes_accession, axis=1, args=(\"ont\",)).values\n",
    "    ))\n",
    "sseq_fields = list(\n",
    "    itt.chain(\n",
    "        *full_table.apply(prefix_split_remotes_accession, axis=1, args=(\"strandseq\",)).values\n",
    "    ))\n",
    "\n",
    "full_table.drop(\n",
    "    [\n",
    "        \"remote_hifi\", \"accession_hifi\",\n",
    "        \"remote_ont\", \"accession_ont\",\n",
    "        \"remote_strandseq\", \"accession_strandseq\"\n",
    "    ], axis=1, inplace=True\n",
    ")\n",
    "\n",
    "column_names = [\"SIN\", \"datatype\", \"remote_path\", \"remote_collection\", \"accession\"]\n",
    "\n",
    "hifi = pd.DataFrame.from_records(hifi_fields, columns=column_names)\n",
    "hifi.set_index(\"SIN\", inplace=True)\n",
    "ont = pd.DataFrame.from_records(ont_fields, columns=column_names)\n",
    "ont.set_index(\"SIN\", inplace=True)\n",
    "sseq = pd.DataFrame.from_records(sseq_fields, columns=column_names)\n",
    "sseq.set_index(\"SIN\", inplace=True)\n",
    "\n",
    "merged = pd.concat([hifi, ont, sseq], axis=0, ignore_index=False)\n",
    "\n",
    "full_table = full_table.merge(merged, left_index=True, right_index=True, how=\"outer\")\n",
    "full_table.sort_values([\"hgsvc_sample_number\", \"datatype\"], inplace=True)\n",
    "\n",
    "with open(out_table, \"w\") as dump:\n",
    "    _ = dump.write(f\"# {TODAY}\\n\")\n",
    "    _ = dump.write(f\"# SOURCE PREFIX PATHS\\n\")\n",
    "    for k, v in prefix_paths.items():\n",
    "        _ = dump.write(f\"# {k} '{v}'\\n\")\n",
    "    full_table.to_csv(dump, sep=\"\\t\", header=True, index=True, index_label=\"SIN\")\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
