{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e64d2d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"../00_project_config.ipynb\"\n",
    "\n",
    "import pathlib as pl\n",
    "import pandas as pd\n",
    "import io\n",
    "\n",
    "local_mount = \"/mounts/hilbert/project\"\n",
    "remote_root = \"/gpfs/project\"\n",
    "\n",
    "BATCH_NUM = 3\n",
    "\n",
    "if BATCH_NUM == 1:\n",
    "    datestamp = \"# 20230818\\n\\n\"\n",
    "    share_root_folder = pl.Path(remote_root).joinpath(\n",
    "        \"projects/medbioinf/data/00_RESTRUCTURE/shares/globus/outgoing/hgsvc/ebi_upload/\"\n",
    "        \"20230818_verkko_batch1\"\n",
    "    )\n",
    "    data_mover_script = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_batch1.sh\"\n",
    "    )\n",
    "    data_mover_pairs = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_pairs_batch1.tsv\"\n",
    "    )\n",
    "elif BATCH_NUM == 2:\n",
    "    datestamp = \"# 20230927\\n\\n\"\n",
    "    share_root_folder = pl.Path(remote_root).joinpath(\n",
    "        \"projects/medbioinf/data/00_RESTRUCTURE/shares/globus/outgoing/hgsvc/ebi_upload/\"\n",
    "        \"20230927_verkko_batch2\"\n",
    "    )\n",
    "    data_mover_script = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_batch2.sh\"\n",
    "    )\n",
    "    data_mover_pairs = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_pairs_batch2.tsv\"\n",
    "    )\n",
    "elif BATCH_NUM == 3:\n",
    "    datestamp = \"# 20240201\\n\\n\"\n",
    "    share_root_folder = pl.Path(remote_root).joinpath(\n",
    "        \"projects/medbioinf/data/00_RESTRUCTURE/shares/globus/outgoing/hgsvc/ebi_upload/\"\n",
    "        \"20240201_verkko_batch3\"\n",
    "    )\n",
    "    data_mover_script = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_batch3.sh\"\n",
    "    )\n",
    "    data_mover_pairs = PROJECT_BASE.joinpath(\n",
    "        \"scripts/datamng\",\n",
    "        \"data_mover_pairs_batch3.tsv\"\n",
    "    )\n",
    "else:\n",
    "    raise\n",
    "\n",
    "folder_specs = [\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/eval/wd/results/statistics/assemblies\",\n",
    "        \"*.summary.tsv\",\n",
    "        \"statistics/assemblies/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/eval/wd/results/reports/ref_chrom_assign\",\n",
    "        \"*.tsv\",\n",
    "        \"reports/ref_chrom_assign/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/eval/wd/results/reports/contamination\",\n",
    "        \"*.norm.tsv\",\n",
    "        \"reports/contamination/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/eval/wd/results/assemblies\",\n",
    "        \"**/*.fasta*\",\n",
    "        \"assemblies/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/eval/wd/results/alignments\",\n",
    "        \"**/*.bam*\",\n",
    "        \"alignments/contig_to_ref/{}/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/verkko/wd/results/reports/seq_dedup\",\n",
    "        \"*ps-sseq*.summary.tsv\",\n",
    "        \"reports/seq_dedup/{}\"\n",
    "    ],\n",
    "    [\n",
    "        \"projects/medbioinf/projects/assemblies/hybrids/verkko/wd/results/assemblies/verkko\",\n",
    "        \"**/*ps-sseq.exemplar*\",\n",
    "        \"assemblies/{}\"\n",
    "    ]\n",
    "]\n",
    "\n",
    "data_mover_script.parent.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "data_mover = io.StringIO()\n",
    "data_mover.write(\"#!/usr/bin/bash\\n\\n\")\n",
    "data_mover.write(datestamp)\n",
    "\n",
    "source_target_table = []\n",
    "for source_folder, source_glob, target_sub in folder_specs:\n",
    "    source_files = sorted(pl.Path(local_mount).joinpath(source_folder).glob(source_glob))\n",
    "    if not source_files:\n",
    "        raise ValueError(source_folder)\n",
    "    \n",
    "    for source_file in source_files:\n",
    "        filename = source_file.name\n",
    "        if \"ps-trio\" in filename or \"ps-hic\" in filename:\n",
    "            continue\n",
    "        if \"karyo-est\" in filename:\n",
    "            continue\n",
    "        sample = source_file.name.split(\".\")[0]\n",
    "        if sample not in HGSVC_SAMPLES[\"sample\"].values:\n",
    "            continue\n",
    "        sample_batch = int(HGSVC_SAMPLES.loc[HGSVC_SAMPLES[\"sample\"] == sample, \"batch_num\"].values[0])\n",
    "        if sample_batch != BATCH_NUM:\n",
    "            continue\n",
    "        if target_sub.startswith(\"alignments\"):\n",
    "            if \"hg38\" in source_file.name:\n",
    "                ref = \"hg38\"\n",
    "            else:\n",
    "                ref = \"t2tv2\"\n",
    "            subfolder = target_sub.format(ref, sample)\n",
    "        else:\n",
    "            subfolder = target_sub.format(sample)\n",
    "        remote_source = str(source_file).replace(local_mount, remote_root)\n",
    "        target_path = share_root_folder.joinpath(subfolder, source_file.name)\n",
    "        source_target_table.append((remote_source, target_path))\n",
    "\n",
    "pairs = pd.DataFrame.from_records(source_target_table, columns=[\"source_path\", \"target_path\"])\n",
    "pairs.to_csv(data_mover_pairs, header=True, index=False, sep=\"\\t\")\n",
    "\n",
    "for row in pairs.itertuples():\n",
    "    target_folder = pl.Path(row.target_path).parent\n",
    "    data_mover.write(\n",
    "        f\"mkdir -p {target_folder}\\n\"\n",
    "    )\n",
    "    data_mover.write(\n",
    "        f\"rsync --progress --checksum {row.source_path} {row.target_path}\\n\"\n",
    "    )\n",
    "    \n",
    "with open(data_mover_script, \"w\") as dump:\n",
    "    dump.write(data_mover.getvalue())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
