{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f69a4dfb-6cd1-449c-ae71-fa0ab68ccb65",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "import pickle as pck\n",
    "\n",
    "%cd -q \"/home/ebertp/work/code/cubi/project-run-hgsvc-hybrid-assemblies/notebooks\"\n",
    "_PROJECT_CONFIG_NB = str(pathlib.Path(\"00_project_config.ipynb\").resolve(strict=True))\n",
    "\n",
    "%run $_PROJECT_CONFIG_NB\n",
    "\n",
    "assembler = \"verkko\"\n",
    "\n",
    "cache_file_chrom_t2t_status = PROJECT_NB_CACHE.joinpath(f\"chrom_t2t_status.{assembler}.pck\")\n",
    "\n",
    "with open(cache_file_chrom_t2t_status, \"rb\") as cache:\n",
    "    data = pck.load(cache)\n",
    "\n",
    "out_table = []\n",
    "\n",
    "t2t_label_remap = {\n",
    "    \"other\": \"not T2T\",\n",
    "    \"t2t_ctg\": \"T2T contig\",\n",
    "    \"t2t_scf\": \"T2T scaffold\"\n",
    "}\n",
    "\n",
    "for sample_key, infos in data.items():\n",
    "    if len(sample_key) == 2:\n",
    "        # just maps sample - sex\n",
    "        continue\n",
    "    sample, hap, chrom = sample_key\n",
    "    sex, contig_id, t2t_label = infos[:3]\n",
    "    remapped_label = t2t_label_remap[t2t_label]\n",
    "    row = (sample, sex, hap, chrom, remapped_label, contig_id)\n",
    "    out_table.append(row)\n",
    "\n",
    "df = pd.DataFrame.from_records(\n",
    "    out_table,\n",
    "    columns=[\"sample\", \"sex\", \"haplotype\", \"chromosome\", \"T2T label\", \"contig ID\"]\n",
    ")\n",
    "\n",
    "def get_chrom_order_number(chrom_name):\n",
    "    try:\n",
    "        num = int(chrom_name.strip(\"chr\"))\n",
    "    except ValueError:\n",
    "        if chrom_name == \"chrX\":\n",
    "            num = 23\n",
    "        elif chrom_name == \"chrY\":\n",
    "            num = 24\n",
    "        else:\n",
    "            raise\n",
    "    return num\n",
    "\n",
    "df[\"chrom_number\"] = df[\"chromosome\"].apply(get_chrom_order_number)\n",
    "df.sort_values([\"sample\", \"haplotype\", \"chrom_number\"], inplace=True)\n",
    "df.drop(\"chrom_number\", axis=1, inplace=True)\n",
    "\n",
    "table_sx_vrk_t2t = TABLE_OUT_SUPPL.joinpath(\n",
    "    f\"table_SXPE_{assembler}_t2t_chrom.tsv\"\n",
    ")\n",
    "\n",
    "df.to_csv(table_sx_vrk_t2t, sep=\"\\t\", header=True, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
