{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f6e4e492",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ast\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pysam\n",
    "import os\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38853370",
   "metadata": {},
   "source": [
    "# Heng Li"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c478ef07",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = '/home/mark/Desktop/HLA/HengLi/Runs/'\n",
    "boundaries={}\n",
    "haplotypes=[]\n",
    "hengCoordinates={}\n",
    "for folder in os.listdir(directory):\n",
    "    boundaries[folder.split(\".\")[0]+\".hap\"+str(folder.split(\".\")[1])]={'Start':0, 'End':0}\n",
    "    haplotypes.append(folder.split(\".\")[0]+\".hap\"+str(folder.split(\".\")[1]))\n",
    "    df = pd.read_csv(directory+str(folder), comment='#', sep='\\t', header=None)\n",
    "    goodRows=[]\n",
    "    for row in df.index:\n",
    "        if 'HLA-DRB' in str(df.at[row,8]):\n",
    "             goodRows.append(row)\n",
    "        else:\n",
    "            continue\n",
    "    df2 = df.loc[goodRows].copy()\n",
    "    df3 = df2[df2[2]=='exon'].copy()\n",
    "    df3[3] = df3[3].astype(int)\n",
    "    df3[4] = df3[4].astype(int)\n",
    "    df3.sort_values(by=[3], inplace=True)\n",
    "    count=len(df3)\n",
    "    start=min(df3[3])\n",
    "    end=max(df3[4])\n",
    "    tempList=[]\n",
    "    for row in df3.index:\n",
    "        tempList.append([x for x in df3.loc[row]])\n",
    "    hengCoordinates[folder.split(\".\")[0]+\".hap\"+str(folder.split(\".\")[1])]={'Start':start, 'End':end, 'Total_Genes':count, 'Exon_List':tempList}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3ec16cc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS CODE WILL GRAB THE BOUNDARIES OF DRA - DQA1\n",
    "directory = '/home/mark/Desktop/HLA/HengLi/Runs/'\n",
    "\n",
    "for folder in os.listdir(directory):\n",
    "\n",
    "    df = pd.read_csv(directory+str(folder), comment='#', sep='\\t', header=None)\n",
    "    goodDRA=[]\n",
    "    goodDQA1=[]\n",
    "    for row in df.index:\n",
    "        if 'HLA-DRA' in str(df.at[row,8]):\n",
    "             goodDRA.append(row)\n",
    "        elif 'HLA-DQA1' in str(df.at[row,8]):\n",
    "             goodDQA1.append(row)\n",
    "        \n",
    "        else:\n",
    "            continue\n",
    "    df2 = df.loc[goodDRA].copy()\n",
    "    df3 = df2[df2[2]=='gene'].copy()\n",
    "    df3[3] = df3[3].astype(int)\n",
    "    df3[4] = df3[4].astype(int)\n",
    "    df3.sort_values(by=[3], inplace=True)\n",
    "    start=max(df3[4])\n",
    "    \n",
    "    df2_dqa1 = df.loc[goodDQA1].copy()\n",
    "    df3_dqa1 = df2_dqa1[df2_dqa1[2]=='gene'].copy()\n",
    "    df3_dqa1[3] = df3_dqa1[3].astype(int)\n",
    "    df3_dqa1[4] = df3_dqa1[4].astype(int)\n",
    "    df3_dqa1.sort_values(by=[3], inplace=True)\n",
    "    end=min(df3_dqa1[3])\n",
    "\n",
    "    hengCoordinates[folder.split(\".\")[0]+\".hap\"+str(folder.split(\".\")[1])]['Start']=start\n",
    "    hengCoordinates[folder.split(\".\")[0]+\".hap\"+str(folder.split(\".\")[1])]['End']=end"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c5828aa",
   "metadata": {},
   "source": [
    "# My data Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "af87d572",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HG02666.hap2',\n",
       " 'HG02059.hap2',\n",
       " 'NA18534.hap1',\n",
       " 'NA19240.hap1',\n",
       " 'NA19317.hap1',\n",
       " 'HG02818.hap1',\n",
       " 'HG03683.hap2',\n",
       " 'NA24385.hap1',\n",
       " 'NA19836.hap1',\n",
       " 'NA21487.hap1',\n",
       " 'HG03683.hap1',\n",
       " 'HG01890.hap1',\n",
       " 'HG03456.hap2',\n",
       " 'chm13.hap1',\n",
       " 'HG02282.hap1',\n",
       " 'HG00268.hap1',\n",
       " 'NA12329.hap2',\n",
       " 'HG00513.hap2',\n",
       " 'HG03371.hap2',\n",
       " 'HG00171.hap1',\n",
       " 'NA19331.hap1',\n",
       " 'NA19705.hap1',\n",
       " 'NA19317.hap2',\n",
       " 'NA12329.hap1',\n",
       " 'NA19434.hap1',\n",
       " 'HG00358.hap2',\n",
       " 'HG00096.hap2',\n",
       " 'HG00732.hap2',\n",
       " 'HG03009.hap1',\n",
       " 'HG01573.hap1',\n",
       " 'HG00171.hap2',\n",
       " 'HG00731.hap1',\n",
       " 'NA19384.hap1',\n",
       " 'HG01457.hap2',\n",
       " 'HG02666.hap1',\n",
       " 'HG00733.hap2',\n",
       " 'HG03371.hap1',\n",
       " 'HG03732.hap2',\n",
       " 'NA19239.hap2',\n",
       " 'NA20355.hap2',\n",
       " 'HG02011.hap2',\n",
       " 'NA19036.hap1',\n",
       " 'NA19836.hap2',\n",
       " 'NA19705.hap2',\n",
       " 'hg38.hap8',\n",
       " 'HG00513.hap1',\n",
       " 'HG01505.hap1',\n",
       " 'HG00512.hap2',\n",
       " 'HG02953.hap2',\n",
       " 'HG03807.hap1',\n",
       " 'HG01457.hap1',\n",
       " 'HG02018.hap1',\n",
       " 'HG02587.hap2',\n",
       " 'NA18534.hap2',\n",
       " 'HG03456.hap1',\n",
       " 'HG03248.hap1',\n",
       " 'NA19347.hap1',\n",
       " 'HG02587.hap1',\n",
       " 'NA19036.hap2',\n",
       " 'NA20509.hap2',\n",
       " 'HG01352.hap1',\n",
       " 'NA19384.hap2',\n",
       " 'HG03065.hap1',\n",
       " 'NA19129.hap2',\n",
       " 'HG02492.hap1',\n",
       " 'hg38.hap3',\n",
       " 'NA19434.hap2',\n",
       " 'HG02106.hap1',\n",
       " 'HG03520.hap2',\n",
       " 'NA20355.hap1',\n",
       " 'HG04217.hap2',\n",
       " 'HG01596.hap1',\n",
       " 'HG02011.hap1',\n",
       " 'NA19238.hap1',\n",
       " 'hg38.hap7',\n",
       " 'HG03520.hap1',\n",
       " 'HG04036.hap2',\n",
       " 'HG01596.hap2',\n",
       " 'hg38.hap2',\n",
       " 'HG01114.hap2',\n",
       " 'NA18989.hap2',\n",
       " 'hg38.hap6',\n",
       " 'NA19347.hap2',\n",
       " 'NA20509.hap1',\n",
       " 'NA19650.hap1',\n",
       " 'HG00514.hap2',\n",
       " 'HG03452.hap2',\n",
       " 'HG00864.hap1',\n",
       " 'HG04036.hap1',\n",
       " 'HG00358.hap1',\n",
       " 'NA18939.hap2',\n",
       " 'HG00512.hap1',\n",
       " 'HG00268.hap2',\n",
       " 'NA24385.hap2',\n",
       " 'HG02059.hap1',\n",
       " 'HG02554.hap1',\n",
       " 'NA19331.hap2',\n",
       " 'HG03452.hap1',\n",
       " 'HG03065.hap2',\n",
       " 'NA19238.hap2',\n",
       " 'HG03807.hap2',\n",
       " 'HG00096.hap1',\n",
       " 'NA19983.hap2',\n",
       " 'NA18939.hap1',\n",
       " 'HG02018.hap2',\n",
       " 'HG02282.hap2',\n",
       " 'NA19129.hap1',\n",
       " 'HG03732.hap1',\n",
       " 'NA21487.hap2',\n",
       " 'HG02769.hap2',\n",
       " 'HG00733.hap1',\n",
       " 'HG00864.hap2',\n",
       " 'HG01352.hap2',\n",
       " 'HG02554.hap2',\n",
       " 'HG00732.hap1',\n",
       " 'HG01114.hap1',\n",
       " 'NA19983.hap1',\n",
       " 'HG02769.hap1',\n",
       " 'HG03248.hap2',\n",
       " 'HG02492.hap2',\n",
       " 'HG03009.hap2',\n",
       " 'NA18989.hap1',\n",
       " 'HG00731.hap2',\n",
       " 'HG04217.hap1',\n",
       " 'NA19239.hap1',\n",
       " 'HG01573.hap2',\n",
       " 'HG02953.hap1',\n",
       " 'NA19650.hap2',\n",
       " 'HG00514.hap1',\n",
       " 'NA19240.hap2',\n",
       " 'hg38.hap1',\n",
       " 'HG01890.hap2',\n",
       " 'HG02106.hap2',\n",
       " 'NA20847.hap1',\n",
       " 'HG02818.hap2',\n",
       " 'NA20847.hap2',\n",
       " 'HG01505.hap2']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c051dddb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137\n"
     ]
    }
   ],
   "source": [
    "#THIS SHOULD SAY 130\n",
    "print(len(haplotypes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76d852f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "repeatDict={}\n",
    "for file in os.listdir('/home/mark/Desktop/HLA/repeatmasker/plain_RepeatMasker/'):\n",
    "    repeatDict[str(file.split(\".\")[0])+\".\"+str(file.split(\".asm-\")[1].split(\".\")[0])]='/home/mark/Desktop/HLA/repeatmasker/plain_RepeatMasker/'+str(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7ddba462",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hg38.hap3\n",
      "hg38.hap5\n",
      "hg38.hap4\n",
      "hg38.hap6\n",
      "hg38.hap7\n",
      "hg38.hap8\n",
      "hg38.hap2\n",
      "hg38.hap1\n"
     ]
    }
   ],
   "source": [
    "for x in repeatDict:\n",
    "    if 'hg38' in str(x):\n",
    "        print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fb9e7591",
   "metadata": {},
   "outputs": [],
   "source": [
    "for haplotype in set(haplotypes):\n",
    "    \n",
    "    drb_rm_list=[]\n",
    "    minimum= int(hengCoordinates[haplotype]['Start'])\n",
    "    maximum= int(hengCoordinates[haplotype]['End'])\n",
    "    \n",
    "    with open(repeatDict[haplotype], 'r') as file:\n",
    "        for line in file:\n",
    "            goodline = ','.join(line.split())\n",
    "            if '.hap' in str(goodline) and '*' not in str(goodline):\n",
    "                startrm = int(goodline.split(\",\")[5])\n",
    "                endrm = int(goodline.split(\",\")[6])\n",
    "\n",
    "                if startrm >= minimum and startrm <= maximum:\n",
    "                    tempList2 = [x for x in goodline.split(\",\")]\n",
    "                    tempList2.append(str(minimum))\n",
    "                    tempList2.append(str(startrm-minimum))\n",
    "                    drb_rm_list.append(tempList2)\n",
    "                else:\n",
    "                    continue\n",
    "                        \n",
    "        finalDRB = pd.DataFrame(data=drb_rm_list, columns=['Score','Divergence','Ins','Del','Hap','Start','End','Length','Orientation', 'Annotation', 'Element_Group', 'pos1', 'pos2', 'pos3','unimportant','Coordinate_Start','Distance_From_Start'])\n",
    "        finalDRB['UniqueName']=[str(finalDRB.at[x,'Hap'])+\"_\"+str(finalDRB.at[x,'Start'])+\"_\"+str(finalDRB.at[x,'Annotation']) for x in finalDRB.index]\n",
    "        finalDRB.set_index(\"UniqueName\", inplace=True)\n",
    "            \n",
    "    file.close()\n",
    "        \n",
    "    drb_rm_list2=[]\n",
    "    \n",
    "    for item in hengCoordinates[haplotype]['Exon_List']:\n",
    "        drb_rm_list2.append([haplotype, item[3], item[4], item[6], minimum, int(item[3])-minimum, item[8].split(\"gene_name\")[1].split(\";\")[0], item[8].split(\"exon_number\")[1].split(\";\")[0]])\n",
    "    \n",
    "    finalDRB2 = pd.DataFrame(data=drb_rm_list2, columns=['Hap','Start','End','Orientation','Coordinate_Start','Distance_From_Start', 'Gene_Name', 'Exon_Number'])\n",
    "    finalDRB2['UniqueName']=[str(finalDRB2.at[x,'Hap'])+\"_\"+str(finalDRB2.at[x,'Start'])+\"_\"+str(finalDRB2.at[x,'Gene_Name']) for x in finalDRB2.index]\n",
    "    finalDRB2.set_index(\"UniqueName\", inplace=True)\n",
    "        \n",
    "    finalDF = pd.concat([finalDRB, finalDRB2])\n",
    "    finalDF['Start']= finalDF['Start'].astype(int)\n",
    "    finalDF['End']= finalDF['End'].astype(int)\n",
    "    finalDF.sort_values(by=['Start'], inplace=True)\n",
    "        \n",
    "    finalDF.to_csv(\"/home/mark/Desktop/HLA/DRB/DataFrames/full/\"+str(haplotype)+\".csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ebc6b965",
   "metadata": {},
   "outputs": [],
   "source": [
    "repeatDict2={x:repeatDict[x].split(\"/plain_RepeatMasker/\")[1].split(\".out\")[0] for x in repeatDict.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fdf6e6b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HG03807.hap2': 'HG03807.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01114.hap1': 'HG01114.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00171.hap2': 'HG00171.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19434.hap2': 'NA19434.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01352.hap2': 'HG01352.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03371.hap2': 'HG03371.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03683.hap1': 'HG03683.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03683.hap2': 'HG03683.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00733.hap1': 'HG00733.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00514.hap2': 'HG00514.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00732.hap1': 'HG00732.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19036.hap1': 'NA19036.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00512.hap1': 'HG00512.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02818.unassigned': 'HG02818.vrk-ps-sseq.asm-unassigned.HLA.cleaned.fasta',\n",
       " 'NA18939.unassigned': 'NA18939.vrk-ps-sseq.asm-unassigned.HLA.cleaned.fasta',\n",
       " 'HG02587.hap1': 'HG02587.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03732.hap1': 'HG03732.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03371.hap1': 'HG03371.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02666.hap1': 'HG02666.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00358.hap2': 'HG00358.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG04036.hap2': 'HG04036.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03456.hap2': 'HG03456.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02818.hap1': 'HG02818.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'hg38.hap3': 'hg38.vrk.asm-hap3.HLA.cleaned.fasta',\n",
       " 'HG01596.hap2': 'HG01596.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02018.hap1': 'HG02018.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19317.hap1': 'NA19317.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02554.hap1': 'HG02554.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA20847.hap1': 'NA20847.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03248.hap1': 'HG03248.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'hg38.hap5': 'hg38.vrk.asm-hap5.HLA.cleaned.fasta',\n",
       " 'HG01596.hap1': 'HG01596.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19129.hap1': 'NA19129.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01890.hap1': 'HG01890.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19836.hap2': 'NA19836.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG04036.hap1': 'HG04036.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01114.hap2': 'HG01114.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA18939.hap1': 'NA18939.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00171.hap1': 'HG00171.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA12329.hap1': 'NA12329.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03452.hap2': 'HG03452.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19836.hap1': 'NA19836.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00732.hap2': 'HG00732.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'hg38.hap4': 'hg38.vrk.asm-hap4.HLA.cleaned.fasta',\n",
       " 'NA20847.hap2': 'NA20847.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00733.hap2': 'HG00733.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03009.hap1': 'HG03009.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00268.hap1': 'HG00268.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02769.hap1': 'HG02769.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'chm13.hap1': 'chm13.vrk.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00731.hap2': 'HG00731.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03732.hap2': 'HG03732.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19347.hap2': 'NA19347.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA24385.hap2': 'NA24385.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00096.hap2': 'HG00096.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19384.hap1': 'NA19384.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03456.hap1': 'HG03456.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19705.hap2': 'NA19705.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03520.hap2': 'HG03520.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19317.hap2': 'NA19317.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00358.hap1': 'HG00358.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'hg38.hap6': 'hg38.vrk.asm-hap6.HLA.cleaned.fasta',\n",
       " 'HG04217.hap1': 'HG04217.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01505.hap2': 'HG01505.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02282.hap2': 'HG02282.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00864.hap2': 'HG00864.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00731.hap1': 'HG00731.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02106.hap1': 'HG02106.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19331.hap1': 'NA19331.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19384.hap2': 'NA19384.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA12329.hap2': 'NA12329.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19238.hap2': 'NA19238.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19240.hap2': 'NA19240.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02554.hap2': 'HG02554.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00512.hap2': 'HG00512.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA24385.hap1': 'NA24385.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA20509.hap2': 'NA20509.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01573.hap1': 'HG01573.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02011.hap2': 'HG02011.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19239.hap1': 'NA19239.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19650.hap1': 'NA19650.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02059.hap1': 'HG02059.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00864.hap1': 'HG00864.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01573.hap2': 'HG01573.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA20509.hap1': 'NA20509.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA18989.hap1': 'NA18989.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02818.hap2': 'HG02818.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'hg38.hap7': 'hg38.vrk.asm-hap7.HLA.cleaned.fasta',\n",
       " 'NA19239.hap2': 'NA19239.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19983.hap1': 'NA19983.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02282.hap1': 'HG02282.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19240.hap1': 'NA19240.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02018.hap2': 'HG02018.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA18989.hap2': 'NA18989.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA21487.hap2': 'NA21487.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03065.hap2': 'HG03065.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19129.hap2': 'NA19129.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03520.hap1': 'HG03520.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03009.hap2': 'HG03009.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19238.hap1': 'NA19238.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG04217.hap2': 'HG04217.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA18534.hap2': 'NA18534.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02769.hap2': 'HG02769.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00513.hap2': 'HG00513.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03807.hap1': 'HG03807.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01890.hap2': 'HG01890.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19331.hap2': 'NA19331.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA18534.hap1': 'NA18534.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00513.hap1': 'HG00513.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02953.hap1': 'HG02953.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02106.unassigned': 'HG02106.vrk-ps-sseq.asm-unassigned.HLA.cleaned.fasta',\n",
       " 'NA20355.hap1': 'NA20355.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA20355.hap2': 'NA20355.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'hg38.hap8': 'hg38.vrk.asm-hap8.HLA.cleaned.fasta',\n",
       " 'hg38.hap2': 'hg38.vrk.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01352.hap1': 'HG01352.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03452.hap1': 'HG03452.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02011.hap1': 'HG02011.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA18939.hap2': 'NA18939.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03683.unassigned': 'HG03683.vrk-ps-sseq.asm-unassigned.HLA.cleaned.fasta',\n",
       " 'NA19347.hap1': 'NA19347.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG00268.hap2': 'HG00268.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19983.hap2': 'NA19983.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA21487.hap1': 'NA21487.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02492.hap1': 'HG02492.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02953.hap2': 'HG02953.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19036.hap2': 'NA19036.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01457.hap1': 'HG01457.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG03065.hap1': 'HG03065.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'hg38.hap1': 'hg38.vrk.asm-hap1.HLA.cleaned.fasta',\n",
       " 'NA19650.hap2': 'NA19650.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19705.hap1': 'NA19705.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02059.hap2': 'HG02059.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02492.hap2': 'HG02492.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG00514.hap1': 'HG00514.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02018.unassigned': 'HG02018.vrk-ps-sseq.asm-unassigned.HLA.cleaned.fasta',\n",
       " 'HG00096.hap1': 'HG00096.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02106.hap2': 'HG02106.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG03248.hap2': 'HG03248.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'NA19434.hap1': 'NA19434.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG01457.hap2': 'HG01457.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG02587.hap2': 'HG02587.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta',\n",
       " 'HG01505.hap1': 'HG01505.vrk-ps-sseq.asm-hap1.HLA.cleaned.fasta',\n",
       " 'HG02666.hap2': 'HG02666.vrk-ps-sseq.asm-hap2.HLA.cleaned.fasta'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeatDict2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a99f6e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "lengthMatch=[]\n",
    "for haplotype in haplotypes:\n",
    "    drb_rm_list=[]\n",
    "    drb_rm_list2=[]\n",
    "    \n",
    "    minimum= int(hengCoordinates[haplotype]['Start'])\n",
    "    maximum= int(hengCoordinates[haplotype]['End'])\n",
    "    \n",
    "    lengthMatch.append([haplotype, maximum-minimum])\n",
    "    \n",
    "    sequence = pysam.faidx('/home/mark/Desktop/HLA/Globus/cleaned/'+str(repeatDict2[haplotype]), haplotype+\":\"+str(minimum)+\"-\"+str(maximum))\n",
    "    #with open('/home/mark/Desktop/HLA/DRB/Sequences/Individual/'+str(haplotype)+\".fasta\", 'a+') as file:\n",
    "    #    file.write(sequence)\n",
    "    #file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb077c31",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
