{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b68b303",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import pysam\n",
    "import re\n",
    "import collections\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1426642f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleDF = pd.read_csv('/home/mark/Desktop/FunctionalGroup/fullSamples/HG00268_subset.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5581cd8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read</th>\n",
       "      <th>contigHap1</th>\n",
       "      <th>locationHap1</th>\n",
       "      <th>contigHap2</th>\n",
       "      <th>locationHap2</th>\n",
       "      <th>flag1</th>\n",
       "      <th>flag2</th>\n",
       "      <th>hap1CIGAR</th>\n",
       "      <th>hap2CIGAR</th>\n",
       "      <th>hap1Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/102171523/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>94567-97865</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77618555-77621853</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2908=6222N88=13870N94=299N206=1S</td>\n",
       "      <td>1S206=299N94=13870N88=6226N2908=</td>\n",
       "      <td>GTAAGATTCCATTTATATGAAGTGTCCAGGGCAGGAAAATCCACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/93126845/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>106263-108294</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77627773-77629804</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>11=1I182=1D85=1D21=1D39=1D110=1D263=2D733N159=...</td>\n",
       "      <td>56=1D35=1D168=1D32=1D439=1D424=1I161=2D733N262...</td>\n",
       "      <td>CACCCCGACGAGGGCGGAGAGCGATGAGGACAACGACGTCGTCATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/132317941/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>116003-118634</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77618176-77620807</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>736=1X1893=</td>\n",
       "      <td>1893=1X736=</td>\n",
       "      <td>AAAACTCTATTAAAACTTGCTACATGTGGTGTTAAGGCTTTTTTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/126157970/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>117047-120001</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77617939-77620893</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1823=1130S</td>\n",
       "      <td>1130S1823=</td>\n",
       "      <td>GTGGATATGAAATTTAATATCCAGAAGAAAAGCACAGAATAAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/3736971/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>125956-129778</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77607034-77610856</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>153=1X1466=1X2198=2S</td>\n",
       "      <td>2S2198=1X1466=1X153=</td>\n",
       "      <td>GCAGCGGTGTAACAAATTTTAATTCAAAATAGACACTCTGAAAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/73730778/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782264</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933980-71936513</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S540=1I28=1D553=902N148=309N142=2102N165=2582...</td>\n",
       "      <td>112=1295N144=4431N163=2828N89=1881N56=1D144=14...</td>\n",
       "      <td>GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/75434545/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782362</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71930435-71933066</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>2S55=3450N152=1295N144=4431N163=2828N89=1881N2...</td>\n",
       "      <td>GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76154490/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782338</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933944-71936551</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>3S25=1I38=1I5=1I31=1I12=1I8=2I28=1I7=1I2=1I19=...</td>\n",
       "      <td>7S3=1X19=1I125=1295N144=4431N163=2828N89=1881N...</td>\n",
       "      <td>GAGGAGGTGTCATGTTTACTTTTTATTTGAGGAGTACAAACTGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76351524/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782298</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933948-71936515</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>1S144=1295N144=4431N163=2828N89=1881N201=1434N...</td>\n",
       "      <td>GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76612054/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782332</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933913-71936514</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>179=1295N144=4431N163=2828N89=1881N201=1434N85...</td>\n",
       "      <td>CGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample                                read          contigHap1  \\\n",
       "0    HG00268  m64119_210712_091919/102171523/ccs  haplotype1-0000001   \n",
       "1    HG00268   m64119_210712_091919/93126845/ccs  haplotype1-0000001   \n",
       "2    HG00268  m64119_210712_091919/132317941/ccs  haplotype1-0000001   \n",
       "3    HG00268  m64119_210712_091919/126157970/ccs  haplotype1-0000001   \n",
       "4    HG00268    m64119_210712_091919/3736971/ccs  haplotype1-0000001   \n",
       "..       ...                                 ...                 ...   \n",
       "995  HG00268   m64119_210712_091919/73730778/ccs  haplotype1-0000001   \n",
       "996  HG00268   m64119_210712_091919/75434545/ccs  haplotype1-0000001   \n",
       "997  HG00268   m64119_210712_091919/76154490/ccs  haplotype1-0000001   \n",
       "998  HG00268   m64119_210712_091919/76351524/ccs  haplotype1-0000001   \n",
       "999  HG00268   m64119_210712_091919/76612054/ccs  haplotype1-0000001   \n",
       "\n",
       "        locationHap1          contigHap2       locationHap2  flag1  flag2  \\\n",
       "0        94567-97865  haplotype2-0000064  77618555-77621853     16      0   \n",
       "1      106263-108294  haplotype2-0000064  77627773-77629804      0     16   \n",
       "2      116003-118634  haplotype2-0000064  77618176-77620807     16      0   \n",
       "3      117047-120001  haplotype2-0000064  77617939-77620893     16      0   \n",
       "4      125956-129778  haplotype2-0000064  77607034-77610856     16      0   \n",
       "..               ...                 ...                ...    ...    ...   \n",
       "995  5779731-5782264  haplotype2-0000064  71933980-71936513     16      0   \n",
       "996  5779731-5782362  haplotype2-0000064  71930435-71933066     16      0   \n",
       "997  5779731-5782338  haplotype2-0000064  71933944-71936551     16      0   \n",
       "998  5779731-5782298  haplotype2-0000064  71933948-71936515     16      0   \n",
       "999  5779731-5782332  haplotype2-0000064  71933913-71936514     16      0   \n",
       "\n",
       "                                             hap1CIGAR  \\\n",
       "0                     2908=6222N88=13870N94=299N206=1S   \n",
       "1    11=1I182=1D85=1D21=1D39=1D110=1D263=2D733N159=...   \n",
       "2                                          736=1X1893=   \n",
       "3                                           1823=1130S   \n",
       "4                                 153=1X1466=1X2198=2S   \n",
       "..                                                 ...   \n",
       "995  1S540=1I28=1D553=902N148=309N142=2102N165=2582...   \n",
       "996  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "997  3S25=1I38=1I5=1I31=1I12=1I8=2I28=1I7=1I2=1I19=...   \n",
       "998  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "999  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "\n",
       "                                             hap2CIGAR  \\\n",
       "0                     1S206=299N94=13870N88=6226N2908=   \n",
       "1    56=1D35=1D168=1D32=1D439=1D424=1I161=2D733N262...   \n",
       "2                                          1893=1X736=   \n",
       "3                                           1130S1823=   \n",
       "4                                 2S2198=1X1466=1X153=   \n",
       "..                                                 ...   \n",
       "995  112=1295N144=4431N163=2828N89=1881N56=1D144=14...   \n",
       "996  2S55=3450N152=1295N144=4431N163=2828N89=1881N2...   \n",
       "997  7S3=1X19=1I125=1295N144=4431N163=2828N89=1881N...   \n",
       "998  1S144=1295N144=4431N163=2828N89=1881N201=1434N...   \n",
       "999  179=1295N144=4431N163=2828N89=1881N201=1434N85...   \n",
       "\n",
       "                                               hap1Seq  \n",
       "0    GTAAGATTCCATTTATATGAAGTGTCCAGGGCAGGAAAATCCACAG...  \n",
       "1    CACCCCGACGAGGGCGGAGAGCGATGAGGACAACGACGTCGTCATC...  \n",
       "2    AAAACTCTATTAAAACTTGCTACATGTGGTGTTAAGGCTTTTTTCT...  \n",
       "3    GTGGATATGAAATTTAATATCCAGAAGAAAAGCACAGAATAAAAAG...  \n",
       "4    GCAGCGGTGTAACAAATTTTAATTCAAAATAGACACTCTGAAAACA...  \n",
       "..                                                 ...  \n",
       "995  GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...  \n",
       "996  GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...  \n",
       "997  GAGGAGGTGTCATGTTTACTTTTTATTTGAGGAGTACAAACTGAGA...  \n",
       "998  GGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...  \n",
       "999  CGAGGTGTCATGTTTACTTTTTATTTAGGAGTACAAACTGAGACAA...  \n",
       "\n",
       "[1000 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4afb9492",
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS IS JUST USED TO CHECK THAT THE SEQUENCES ARE CLEAN\n",
    "#good = ['A', 'C', 'T', 'G']\n",
    "#for x in sampleDF['hap1Seq']:\n",
    "#    for y in [a for a in x]:\n",
    "#        if str(y.upper()) in good:\n",
    "#            continue\n",
    "#        else:\n",
    "#            print(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e6a3163f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def basicPhaser2(inputCigarString):\n",
    "    matches = re.findall(r'(\\d+)([A-Z=]{1})', inputCigarString)\n",
    "    tempDict={}\n",
    "    total=0\n",
    "    for cig in matches:\n",
    "        if str(cig[1]) in tempDict.keys():\n",
    "            tempDict[cig[1]]+=int(cig[0])\n",
    "        else:\n",
    "            tempDict[cig[1]]=int(cig[0])\n",
    "            \n",
    "    return(tempDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b9d84749",
   "metadata": {},
   "outputs": [],
   "source": [
    "h1 = []\n",
    "h2 = []\n",
    "for x,y in zip(sampleDF['hap1CIGAR'], sampleDF['hap2CIGAR']):\n",
    "    h1.append(basicPhaser2(str(x)))\n",
    "    h2.append(basicPhaser2(str(y)))\n",
    "sampleDF['BasicPhase_H1']=h1\n",
    "sampleDF['BasicPhase_H2']=h2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "439fb810",
   "metadata": {},
   "outputs": [],
   "source": [
    "hap1startSites=[]\n",
    "hap2startSites=[]\n",
    "\n",
    "for row in sampleDF.index:\n",
    "    if str(sampleDF.at[row,'hap1CIGAR']) == 'nan' or str(sampleDF.at[row,'hap2CIGAR']) == 'nan':\n",
    "        hap1startSites.append(\"PHASEDALREADY\")\n",
    "        hap2startSites.append(\"PHASEDALREADY\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        hap1matches = re.findall(r'(\\d+)([A-Z=]{1})', str(sampleDF.at[row,'hap1CIGAR']))\n",
    "        hap2matches = re.findall(r'(\\d+)([A-Z=]{1})', str(sampleDF.at[row,'hap2CIGAR']))\n",
    "        cig1 = hap1matches[0]\n",
    "        cig2 = hap2matches[0]\n",
    "\n",
    "        if str(cig1[1]) == 'S':\n",
    "            hap1startSites.append(int(sampleDF.at[row,'locationHap1'].split(\"-\")[0])-int(cig1[0]))\n",
    "        else:\n",
    "            hap1startSites.append(int(sampleDF.at[row,'locationHap1'].split(\"-\")[0]))\n",
    "\n",
    "        if str(cig2[1]) == 'S':\n",
    "            hap2startSites.append(int(sampleDF.at[row,'locationHap2'].split(\"-\")[0])-int(cig2[0]))\n",
    "        else:\n",
    "            hap2startSites.append(int(sampleDF.at[row,'locationHap2'].split(\"-\")[0]))\n",
    "sampleDF['Hap1Start']=hap1startSites\n",
    "sampleDF['Hap2Start']=hap2startSites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc957d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pullSequenceKmer(s, genome, contig, startSite, mywindow,fliporNo):\n",
    "    \n",
    "    matches = re.findall(r'(\\d+)([A-Z=]{1})', s)\n",
    "    currentlocation = int(startSite)\n",
    "    sequences=[]\n",
    "    \n",
    "    for cig in matches:\n",
    "    \n",
    "        if str(cig[1]) != 'N' and str(cig[1]) != 'S' and str(cig[1]) != 'I':\n",
    "            goTolocation = currentlocation + (int(cig[0])-1)\n",
    "            sequences.append(pysam.faidx(genome, str(contig)+\":\"+str(currentlocation)+\"-\"+str(goTolocation)).split()[1:])\n",
    "            currentlocation=goTolocation+1\n",
    "            \n",
    "        elif str(cig[1]) == 'I':\n",
    "            continue\n",
    "        \n",
    "        else:\n",
    "            currentlocation = (currentlocation + (int(cig[0])-1))+1\n",
    "                \n",
    "    newSequences = []\n",
    "    for x in sequences:\n",
    "        for y in x:\n",
    "            newSequences.append(y)\n",
    "           \n",
    "    mySequence = ''.join(newSequences)\n",
    "    del newSequences, sequences\n",
    "    \n",
    "    if fliporNo.upper() == 'NO':\n",
    "        mykmers = []\n",
    "        i =0\n",
    "        while i < (len(mySequence)-mywindow)+1:\n",
    "            mykmers.append(mySequence[i:i+mywindow])\n",
    "            i+=1\n",
    "    else:\n",
    "        mySequence2 = Seq(mySequence)\n",
    "        mySequence3 = str(mySequence2.reverse_complement())\n",
    "        mykmers = []\n",
    "        i =0\n",
    "        while i < (len(mySequence3)-mywindow)+1:\n",
    "            mykmers.append(mySequence3[i:i+mywindow])\n",
    "            i+=1\n",
    "    \n",
    "    \n",
    "    return(mykmers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f2be8839",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 1000/1000 [01:53<00:00,  8.84it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "missingHap1 = []\n",
    "missingHap2 = []\n",
    "for row in tqdm(sampleDF.index):\n",
    "    \n",
    "    if sampleDF.at[row,'Hap1Start'] =='PHASEDALREADY':\n",
    "        missingHap1.append([])\n",
    "        missingHap2.append([])\n",
    "        \n",
    "    else:\n",
    "        if str(sampleDF.at[row,'flag1']) == str(sampleDF.at[row,'flag2']):\n",
    "            fliporNoinput = 'NO'\n",
    "        else:\n",
    "            fliporNoinput = 'YES'\n",
    "\n",
    "        myRead = str(sampleDF.at[row,'hap1Seq'].strip())\n",
    "        myReadKmers=[]\n",
    "        window = 15\n",
    "        i =0\n",
    "        while i < (len(myRead)-window)+1:\n",
    "            myReadKmers.append(myRead[i:i+window])\n",
    "            i+=1\n",
    "\n",
    "\n",
    "        hap1SeqKmer = pullSequenceKmer(sampleDF.at[row,'hap1CIGAR'],'/home/mark/Desktop/FunctionalGroup/genomes/HG00268.vrk-ps-sseq.asm-hap1.fasta',str(sampleDF.at[row,'contigHap1']),int(sampleDF.at[row,'Hap1Start']),15, 'NO')\n",
    "        hap2SeqKmer = pullSequenceKmer(sampleDF.at[row,'hap2CIGAR'],'/home/mark/Desktop/FunctionalGroup/genomes/HG00268.vrk-ps-sseq.asm-hap2.fasta',str(sampleDF.at[row,'contigHap2']),int(sampleDF.at[row,'Hap2Start']),15, fliporNoinput)\n",
    "\n",
    "        mhap1 = []\n",
    "        mhap2 = []\n",
    "        for kmer in myReadKmers:\n",
    "            if kmer in hap1SeqKmer:\n",
    "                pass\n",
    "            else:\n",
    "                mhap1.append(kmer)\n",
    "\n",
    "            if kmer in hap2SeqKmer:\n",
    "                pass\n",
    "            else:\n",
    "                mhap2.append(kmer)\n",
    "        missingHap1.append(mhap1)\n",
    "        missingHap2.append(mhap2)\n",
    "sampleDF['Hap1Missing']=missingHap1\n",
    "sampleDF['Hap2Missing']=missingHap2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "42cada17",
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleDF['PhaseV1Results']='temp'\n",
    "sampleDF['PhaseV2Results']='temp'\n",
    "\n",
    "for row in sampleDF.index:\n",
    "    if int(sampleDF.at[row,'BasicPhase_H1']['='])>int(sampleDF.at[row,'BasicPhase_H2']['=']):\n",
    "        sampleDF.at[row,'PhaseV1Results']='Haplotype1'\n",
    "    elif int(sampleDF.at[row,'BasicPhase_H1']['='])<int(sampleDF.at[row,'BasicPhase_H2']['=']):\n",
    "        sampleDF.at[row,'PhaseV1Results']='Haplotype2'\n",
    "    \n",
    "    else:\n",
    "        if sampleDF.at[row,'Hap1Start'] =='PHASEDALREADY':\n",
    "            if str(sampleDF.at[row,'hap1CIGAR']) == 'nan' and str(sampleDF.at[row,'hap2CIGAR']) != 'nan':\n",
    "                sampleDF.at[row,'PhaseV1Results']='Haplotype2'\n",
    "            else:\n",
    "                sampleDF.at[row,'PhaseV1Results']='Haplotype1'\n",
    "        else:\n",
    "            sampleDF.at[row,'PhaseV1Results']='NotAbleToPhase'\n",
    "        \n",
    "    if len(sampleDF.at[row,'Hap1Missing'])<len(sampleDF.at[row,'Hap2Missing']):\n",
    "        sampleDF.at[row,'PhaseV2Results']='Haplotype1'\n",
    "    elif len(sampleDF.at[row,'Hap1Missing'])>len(sampleDF.at[row,'Hap2Missing']):\n",
    "        sampleDF.at[row,'PhaseV2Results']='Haplotype2'\n",
    "    \n",
    "    else:\n",
    "        if sampleDF.at[row,'Hap1Start'] =='PHASEDALREADY':\n",
    "            if str(sampleDF.at[row,'hap1CIGAR']) == 'nan' and str(sampleDF.at[row,'hap2CIGAR']) != 'nan':\n",
    "                sampleDF.at[row,'PhaseV2Results']='Haplotype2'\n",
    "            else:\n",
    "                sampleDF.at[row,'PhaseV2Results']='Haplotype1'\n",
    "        else:\n",
    "            sampleDF.at[row,'PhaseV2Results']='NotAbleToPhase'    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d9e116e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "deepCigar=[]\n",
    "for row in sampleDF.index:\n",
    "    total1=0\n",
    "    total2=0\n",
    "    \n",
    "    h1Dict = sampleDF.at[row,'BasicPhase_H1']\n",
    "    h2Dict = sampleDF.at[row,'BasicPhase_H2']\n",
    "    \n",
    "    for key in h1Dict.keys():\n",
    "        if key == '=':\n",
    "            total1+=int(h1Dict[key])\n",
    "        elif key == 'I':\n",
    "            total1-=int(h1Dict[key])\n",
    "        elif key == 'S':\n",
    "            total1-=int(h1Dict[key])\n",
    "            \n",
    "        elif key == 'N':\n",
    "            continue\n",
    "            \n",
    "        elif key == 'D':\n",
    "            total1-=int(h1Dict[key])\n",
    "            \n",
    "        elif key == 'X':\n",
    "            total1-=int(h1Dict[key])\n",
    "        \n",
    "        else:\n",
    "            continue\n",
    "    for key in h2Dict.keys():\n",
    "        if key == '=':\n",
    "            total2+=int(h2Dict[key])\n",
    "        elif key == 'I':\n",
    "            total2-=int(h2Dict[key])\n",
    "        elif key == 'S':\n",
    "            total2-=int(h2Dict[key])\n",
    "            \n",
    "        elif key == 'N':\n",
    "            continue\n",
    "            \n",
    "        elif key == 'D':\n",
    "            total2-=int(h2Dict[key])\n",
    "            \n",
    "        elif key == 'X':\n",
    "            total2-=int(h2Dict[key])\n",
    "        \n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "    if total1 > total2:\n",
    "        deepCigar.append(\"Haplotype1\")\n",
    "    elif total2> total1:\n",
    "        deepCigar.append(\"Haplotype2\")\n",
    "    else:\n",
    "        deepCigar.append(\"NotAbleToPhase\")\n",
    "sampleDF['PhaseV1_DeepResults']=deepCigar            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b162348",
   "metadata": {},
   "outputs": [],
   "source": [
    "FinalPhaseList=[]\n",
    "for row in sampleDF.index:\n",
    "    tempList =[]\n",
    "    tempList.append(sampleDF.at[row,'PhaseV1Results'])\n",
    "    tempList.append(sampleDF.at[row,'PhaseV2Results'])\n",
    "    tempList.append(sampleDF.at[row,'PhaseV1_DeepResults'])\n",
    "    #print(tempList)\n",
    "    tempDict = collections.Counter(tempList)\n",
    "    #print(tempDict)\n",
    "    maxPhase = max(tempDict, key=tempDict.get)\n",
    "    if int(tempDict[maxPhase]) == 1:\n",
    "        FinalPhaseList.append('NotAbleToPhase')\n",
    "    else:\n",
    "        FinalPhaseList.append(maxPhase)\n",
    "sampleDF['Final_Phase']=FinalPhaseList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "12194d8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read</th>\n",
       "      <th>contigHap1</th>\n",
       "      <th>locationHap1</th>\n",
       "      <th>contigHap2</th>\n",
       "      <th>locationHap2</th>\n",
       "      <th>flag1</th>\n",
       "      <th>flag2</th>\n",
       "      <th>hap1CIGAR</th>\n",
       "      <th>hap2CIGAR</th>\n",
       "      <th>...</th>\n",
       "      <th>BasicPhase_H1</th>\n",
       "      <th>BasicPhase_H2</th>\n",
       "      <th>Hap1Start</th>\n",
       "      <th>Hap2Start</th>\n",
       "      <th>Hap1Missing</th>\n",
       "      <th>Hap2Missing</th>\n",
       "      <th>PhaseV1Results</th>\n",
       "      <th>PhaseV2Results</th>\n",
       "      <th>PhaseV1_DeepResults</th>\n",
       "      <th>Final_Phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/102171523/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>94567-97865</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77618555-77621853</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2908=6222N88=13870N94=299N206=1S</td>\n",
       "      <td>1S206=299N94=13870N88=6226N2908=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3296, 'N': 20391, 'S': 1}</td>\n",
       "      <td>{'S': 1, '=': 3296, 'N': 20395}</td>\n",
       "      <td>94567</td>\n",
       "      <td>77618554</td>\n",
       "      <td>[CCAGGGCAGCAGCTC]</td>\n",
       "      <td>[CCAGGGCAGCAGCTC]</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/93126845/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>106263-108294</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77627773-77629804</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>11=1I182=1D85=1D21=1D39=1D110=1D263=2D733N159=...</td>\n",
       "      <td>56=1D35=1D168=1D32=1D439=1D424=1I161=2D733N262...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2028, 'I': 2, 'D': 12, 'N': 733}</td>\n",
       "      <td>{'=': 2028, 'D': 12, 'I': 2, 'N': 733}</td>\n",
       "      <td>106263</td>\n",
       "      <td>77627773</td>\n",
       "      <td>[CACCCCGACGAGGGC, ACCCCGACGAGGGCG, CCCCGACGAGG...</td>\n",
       "      <td>[CACCCCGACGAGGGC, ACCCCGACGAGGGCG, CCCCGACGAGG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/132317941/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>116003-118634</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77618176-77620807</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>736=1X1893=</td>\n",
       "      <td>1893=1X736=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2629, 'X': 1}</td>\n",
       "      <td>{'=': 2629, 'X': 1}</td>\n",
       "      <td>116003</td>\n",
       "      <td>77618176</td>\n",
       "      <td>[CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...</td>\n",
       "      <td>[CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/126157970/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>117047-120001</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77617939-77620893</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1823=1130S</td>\n",
       "      <td>1130S1823=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 1823, 'S': 1130}</td>\n",
       "      <td>{'S': 1130, '=': 1823}</td>\n",
       "      <td>117047</td>\n",
       "      <td>77616809</td>\n",
       "      <td>[CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...</td>\n",
       "      <td>[CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/3736971/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>125956-129778</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77607034-77610856</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>153=1X1466=1X2198=2S</td>\n",
       "      <td>2S2198=1X1466=1X153=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3817, 'X': 2, 'S': 2}</td>\n",
       "      <td>{'S': 2, '=': 3817, 'X': 2}</td>\n",
       "      <td>125956</td>\n",
       "      <td>77607032</td>\n",
       "      <td>[ATGAACATTATAGCG, TGAACATTATAGCGA, GAACATTATAG...</td>\n",
       "      <td>[ATGAACATTATAGCG, TGAACATTATAGCGA, GAACATTATAG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/73730778/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782264</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933980-71936513</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S540=1I28=1D553=902N148=309N142=2102N165=2582...</td>\n",
       "      <td>112=1295N144=4431N163=2828N89=1881N56=1D144=14...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 1, '=': 2529, 'I': 2, 'D': 3, 'N': 18794}</td>\n",
       "      <td>{'=': 2529, 'N': 18621, 'D': 3, 'I': 2, 'S': 1}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71933980</td>\n",
       "      <td>[GGAGGTGTCATGTTT, GCTCAAACTCTTGGG, CTCAAACTCTT...</td>\n",
       "      <td>[GGAGGTGTCATGTTT, GCTCAAACTCTTGGG, CTCAAACTCTT...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/75434545/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782362</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71930435-71933066</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>2S55=3450N152=1295N144=4431N163=2828N89=1881N2...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 3, '=': 2625, 'N': 22244, 'X': 2}</td>\n",
       "      <td>{'S': 3, '=': 2627, 'N': 22071}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71930433</td>\n",
       "      <td>[GGAGGTGTCATGTTT, CTAGTGGAGACTGAG, TAGTGGAGACT...</td>\n",
       "      <td>[GGAGGTGTCATGTTT, GCAAGCGCGGCGCCC, CAAGCGCGGCG...</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76154490/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782338</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933944-71936551</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>3S25=1I38=1I5=1I31=1I12=1I8=2I28=1I7=1I2=1I19=...</td>\n",
       "      <td>7S3=1X19=1I125=1295N144=4431N163=2828N89=1881N...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 7, '=': 2568, 'I': 31, 'D': 3, 'N': 18794}</td>\n",
       "      <td>{'S': 10, '=': 2564, 'X': 1, 'I': 31, 'N': 186...</td>\n",
       "      <td>5779728</td>\n",
       "      <td>71933937</td>\n",
       "      <td>[GAGGAGGTGTCATGT, AGGAGGTGTCATGTT, GGAGGTGTCAT...</td>\n",
       "      <td>[GAGGAGGTGTCATGT, AGGAGGTGTCATGTT, GGAGGTGTCAT...</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76351524/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782298</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933948-71936515</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>1S144=1295N144=4431N163=2828N89=1881N201=1434N...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 1, '=': 2565, 'N': 18794}</td>\n",
       "      <td>{'S': 2, '=': 2564, 'N': 18621}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71933947</td>\n",
       "      <td>[GGAGGTGTCATGTTT]</td>\n",
       "      <td>[GGAGGTGTCATGTTT, CTAGTGGAGACTGAC]</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76612054/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782332</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933913-71936514</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S1122=902N148=309N142=2102N165=2582N161=1026N...</td>\n",
       "      <td>179=1295N144=4431N163=2828N89=1881N201=1434N85...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 1, '=': 2597, 'N': 18794, 'X': 2}</td>\n",
       "      <td>{'=': 2599, 'N': 18621, 'S': 1}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71933913</td>\n",
       "      <td>[CGAGGTGTCATGTTT, CTAGTGGAGACTGAG, TAGTGGAGACT...</td>\n",
       "      <td>[CGAGGTGTCATGTTT]</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample                                read          contigHap1  \\\n",
       "0    HG00268  m64119_210712_091919/102171523/ccs  haplotype1-0000001   \n",
       "1    HG00268   m64119_210712_091919/93126845/ccs  haplotype1-0000001   \n",
       "2    HG00268  m64119_210712_091919/132317941/ccs  haplotype1-0000001   \n",
       "3    HG00268  m64119_210712_091919/126157970/ccs  haplotype1-0000001   \n",
       "4    HG00268    m64119_210712_091919/3736971/ccs  haplotype1-0000001   \n",
       "..       ...                                 ...                 ...   \n",
       "995  HG00268   m64119_210712_091919/73730778/ccs  haplotype1-0000001   \n",
       "996  HG00268   m64119_210712_091919/75434545/ccs  haplotype1-0000001   \n",
       "997  HG00268   m64119_210712_091919/76154490/ccs  haplotype1-0000001   \n",
       "998  HG00268   m64119_210712_091919/76351524/ccs  haplotype1-0000001   \n",
       "999  HG00268   m64119_210712_091919/76612054/ccs  haplotype1-0000001   \n",
       "\n",
       "        locationHap1          contigHap2       locationHap2  flag1  flag2  \\\n",
       "0        94567-97865  haplotype2-0000064  77618555-77621853     16      0   \n",
       "1      106263-108294  haplotype2-0000064  77627773-77629804      0     16   \n",
       "2      116003-118634  haplotype2-0000064  77618176-77620807     16      0   \n",
       "3      117047-120001  haplotype2-0000064  77617939-77620893     16      0   \n",
       "4      125956-129778  haplotype2-0000064  77607034-77610856     16      0   \n",
       "..               ...                 ...                ...    ...    ...   \n",
       "995  5779731-5782264  haplotype2-0000064  71933980-71936513     16      0   \n",
       "996  5779731-5782362  haplotype2-0000064  71930435-71933066     16      0   \n",
       "997  5779731-5782338  haplotype2-0000064  71933944-71936551     16      0   \n",
       "998  5779731-5782298  haplotype2-0000064  71933948-71936515     16      0   \n",
       "999  5779731-5782332  haplotype2-0000064  71933913-71936514     16      0   \n",
       "\n",
       "                                             hap1CIGAR  \\\n",
       "0                     2908=6222N88=13870N94=299N206=1S   \n",
       "1    11=1I182=1D85=1D21=1D39=1D110=1D263=2D733N159=...   \n",
       "2                                          736=1X1893=   \n",
       "3                                           1823=1130S   \n",
       "4                                 153=1X1466=1X2198=2S   \n",
       "..                                                 ...   \n",
       "995  1S540=1I28=1D553=902N148=309N142=2102N165=2582...   \n",
       "996  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "997  3S25=1I38=1I5=1I31=1I12=1I8=2I28=1I7=1I2=1I19=...   \n",
       "998  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "999  1S1122=902N148=309N142=2102N165=2582N161=1026N...   \n",
       "\n",
       "                                             hap2CIGAR  ...  \\\n",
       "0                     1S206=299N94=13870N88=6226N2908=  ...   \n",
       "1    56=1D35=1D168=1D32=1D439=1D424=1I161=2D733N262...  ...   \n",
       "2                                          1893=1X736=  ...   \n",
       "3                                           1130S1823=  ...   \n",
       "4                                 2S2198=1X1466=1X153=  ...   \n",
       "..                                                 ...  ...   \n",
       "995  112=1295N144=4431N163=2828N89=1881N56=1D144=14...  ...   \n",
       "996  2S55=3450N152=1295N144=4431N163=2828N89=1881N2...  ...   \n",
       "997  7S3=1X19=1I125=1295N144=4431N163=2828N89=1881N...  ...   \n",
       "998  1S144=1295N144=4431N163=2828N89=1881N201=1434N...  ...   \n",
       "999  179=1295N144=4431N163=2828N89=1881N201=1434N85...  ...   \n",
       "\n",
       "                                        BasicPhase_H1  \\\n",
       "0                     {'=': 3296, 'N': 20391, 'S': 1}   \n",
       "1              {'=': 2028, 'I': 2, 'D': 12, 'N': 733}   \n",
       "2                                 {'=': 2629, 'X': 1}   \n",
       "3                              {'=': 1823, 'S': 1130}   \n",
       "4                         {'=': 3817, 'X': 2, 'S': 2}   \n",
       "..                                                ...   \n",
       "995   {'S': 1, '=': 2529, 'I': 2, 'D': 3, 'N': 18794}   \n",
       "996           {'S': 3, '=': 2625, 'N': 22244, 'X': 2}   \n",
       "997  {'S': 7, '=': 2568, 'I': 31, 'D': 3, 'N': 18794}   \n",
       "998                   {'S': 1, '=': 2565, 'N': 18794}   \n",
       "999           {'S': 1, '=': 2597, 'N': 18794, 'X': 2}   \n",
       "\n",
       "                                         BasicPhase_H2 Hap1Start  Hap2Start  \\\n",
       "0                      {'S': 1, '=': 3296, 'N': 20395}     94567   77618554   \n",
       "1               {'=': 2028, 'D': 12, 'I': 2, 'N': 733}    106263   77627773   \n",
       "2                                  {'=': 2629, 'X': 1}    116003   77618176   \n",
       "3                               {'S': 1130, '=': 1823}    117047   77616809   \n",
       "4                          {'S': 2, '=': 3817, 'X': 2}    125956   77607032   \n",
       "..                                                 ...       ...        ...   \n",
       "995    {'=': 2529, 'N': 18621, 'D': 3, 'I': 2, 'S': 1}   5779730   71933980   \n",
       "996                    {'S': 3, '=': 2627, 'N': 22071}   5779730   71930433   \n",
       "997  {'S': 10, '=': 2564, 'X': 1, 'I': 31, 'N': 186...   5779728   71933937   \n",
       "998                    {'S': 2, '=': 2564, 'N': 18621}   5779730   71933947   \n",
       "999                    {'=': 2599, 'N': 18621, 'S': 1}   5779730   71933913   \n",
       "\n",
       "                                           Hap1Missing  \\\n",
       "0                                    [CCAGGGCAGCAGCTC]   \n",
       "1    [CACCCCGACGAGGGC, ACCCCGACGAGGGCG, CCCCGACGAGG...   \n",
       "2    [CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...   \n",
       "3    [CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...   \n",
       "4    [ATGAACATTATAGCG, TGAACATTATAGCGA, GAACATTATAG...   \n",
       "..                                                 ...   \n",
       "995  [GGAGGTGTCATGTTT, GCTCAAACTCTTGGG, CTCAAACTCTT...   \n",
       "996  [GGAGGTGTCATGTTT, CTAGTGGAGACTGAG, TAGTGGAGACT...   \n",
       "997  [GAGGAGGTGTCATGT, AGGAGGTGTCATGTT, GGAGGTGTCAT...   \n",
       "998                                  [GGAGGTGTCATGTTT]   \n",
       "999  [CGAGGTGTCATGTTT, CTAGTGGAGACTGAG, TAGTGGAGACT...   \n",
       "\n",
       "                                           Hap2Missing  PhaseV1Results  \\\n",
       "0                                    [CCAGGGCAGCAGCTC]  NotAbleToPhase   \n",
       "1    [CACCCCGACGAGGGC, ACCCCGACGAGGGCG, CCCCGACGAGG...  NotAbleToPhase   \n",
       "2    [CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...  NotAbleToPhase   \n",
       "3    [CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...  NotAbleToPhase   \n",
       "4    [ATGAACATTATAGCG, TGAACATTATAGCGA, GAACATTATAG...  NotAbleToPhase   \n",
       "..                                                 ...             ...   \n",
       "995  [GGAGGTGTCATGTTT, GCTCAAACTCTTGGG, CTCAAACTCTT...  NotAbleToPhase   \n",
       "996  [GGAGGTGTCATGTTT, GCAAGCGCGGCGCCC, CAAGCGCGGCG...      Haplotype2   \n",
       "997  [GAGGAGGTGTCATGT, AGGAGGTGTCATGTT, GGAGGTGTCAT...      Haplotype1   \n",
       "998                 [GGAGGTGTCATGTTT, CTAGTGGAGACTGAC]      Haplotype1   \n",
       "999                                  [CGAGGTGTCATGTTT]      Haplotype2   \n",
       "\n",
       "     PhaseV2Results PhaseV1_DeepResults     Final_Phase  \n",
       "0    NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "1    NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "2    NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "3    NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "4    NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "..              ...                 ...             ...  \n",
       "995  NotAbleToPhase      NotAbleToPhase  NotAbleToPhase  \n",
       "996      Haplotype2          Haplotype2      Haplotype2  \n",
       "997      Haplotype1          Haplotype1      Haplotype1  \n",
       "998      Haplotype1          Haplotype1      Haplotype1  \n",
       "999      Haplotype2          Haplotype2      Haplotype2  \n",
       "\n",
       "[1000 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f4ddcb65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'Haplotype2': 403, 'Haplotype1': 317, 'NotAbleToPhase': 280})\n",
      "Counter({'Haplotype1': 417, 'Haplotype2': 404, 'NotAbleToPhase': 179})\n",
      "Counter({'Haplotype1': 421, 'Haplotype2': 403, 'NotAbleToPhase': 176})\n",
      "Counter({'Haplotype1': 413, 'Haplotype2': 402, 'NotAbleToPhase': 185})\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "print(collections.Counter(sampleDF['PhaseV1Results']))\n",
    "print(collections.Counter(sampleDF['PhaseV1_DeepResults']))\n",
    "print(collections.Counter(sampleDF['PhaseV2Results']))\n",
    "print(collections.Counter(sampleDF['Final_Phase']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a1d8cb4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read</th>\n",
       "      <th>contigHap1</th>\n",
       "      <th>locationHap1</th>\n",
       "      <th>contigHap2</th>\n",
       "      <th>locationHap2</th>\n",
       "      <th>flag1</th>\n",
       "      <th>flag2</th>\n",
       "      <th>hap1CIGAR</th>\n",
       "      <th>hap2CIGAR</th>\n",
       "      <th>...</th>\n",
       "      <th>BasicPhase_H2</th>\n",
       "      <th>Hap1Start</th>\n",
       "      <th>Hap2Start</th>\n",
       "      <th>Hap1Missing</th>\n",
       "      <th>Hap2Missing</th>\n",
       "      <th>PhaseV1Results</th>\n",
       "      <th>PhaseV2Results</th>\n",
       "      <th>Breakdown</th>\n",
       "      <th>PhaseV1_DeepResults</th>\n",
       "      <th>Final_Phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/145165136/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>293599-295689</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77437144-77439234</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1S373=2255N1018=2I553=1825N142=</td>\n",
       "      <td>81=1X60=1825N536=1I1036=2243N373=1S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2086, 'X': 1, 'N': 4068, 'I': 1, 'S': 1}</td>\n",
       "      <td>293598</td>\n",
       "      <td>77437144</td>\n",
       "      <td>[GACATTTTTAAGTTT]</td>\n",
       "      <td>[GACATTTTTAAGTTT, TCCATGTGGACCGCG, CCATGTGGACC...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/29230379/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>3166242-3171367</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>74450787-74455912</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>479=26221N76=39010N88=48869N349=1I494=1I260=1I...</td>\n",
       "      <td>436=1I525=1I493=1I1017=1I888=1I266=1I497=1I352...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 5116, 'I': 7, 'N': 114023, 'X': 1}</td>\n",
       "      <td>3166242</td>\n",
       "      <td>74450787</td>\n",
       "      <td>[CATGGCTTCCTCCCC, ATGGCTTCCTCCCCA, TGGCTTCCTCC...</td>\n",
       "      <td>[CCAGGAGAGCCCGGA, CAGGAGAGCCCGGAT, AGGAGAGCCCG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/64226014/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5090954-5094461</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72635129-72638636</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1S310=6509N1100=1I2094=</td>\n",
       "      <td>2095=1I1099=6509N310=1S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3504, 'I': 1, 'N': 6509, 'S': 1}</td>\n",
       "      <td>5090953</td>\n",
       "      <td>72635129</td>\n",
       "      <td>[GCCCGCCCGCCCTAG, GGGGAAAAAAAAAAA, GGGAAAAAAAA...</td>\n",
       "      <td>[GCCCGCCCGCCCTAG, TCTTCCCTTGGGGGG, CTTCCCTTGGG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/120588203/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5090974-5093553</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72636037-72638616</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1S290=6509N1100=1I1186=</td>\n",
       "      <td>1187=1I1099=6509N290=1S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2576, 'I': 1, 'N': 6509, 'S': 1}</td>\n",
       "      <td>5090973</td>\n",
       "      <td>72636037</td>\n",
       "      <td>[GAGAGCACAGCCGAG, GGGGAAAAAAAAAAA, GGGAAAAAAAA...</td>\n",
       "      <td>[GAGAGCACAGCCGAG, TCTTCCCTTGGGGGG, CTTCCCTTGGG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/126222455/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5090979-5095855</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72633734-72638610</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>285=6509N7=1D1092=1I3489=1S</td>\n",
       "      <td>1S3490=1I1090=1D8=6509N285=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 1, '=': 4873, 'I': 1, 'D': 1, 'N': 6509}</td>\n",
       "      <td>5090979</td>\n",
       "      <td>72633733</td>\n",
       "      <td>[GAACCGATTTGTTGT, AACCGATTTGTTGTG, ACCGATTTGTT...</td>\n",
       "      <td>[GAACCGATTTGTTGT, AACCGATTTGTTGTG, ACCGATTTGTT...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/26346562/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5091010-5094148</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72635447-72638585</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>2S132=1I122=6509N337=1I530=1I233=1I364=1I1412=</td>\n",
       "      <td>1409=1I368=1I226=1I531=1I342=6509N118=1I136=2S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3130, 'I': 5, 'N': 6509, 'S': 2}</td>\n",
       "      <td>5091008</td>\n",
       "      <td>72635447</td>\n",
       "      <td>[GGGACCCCGGCCAGC, GGACCCCGGCCAGCG, CTCCGCCATTC...</td>\n",
       "      <td>[GGGACCCCGGCCAGC, GGACCCCGGCCAGCG, CTCCGCCATTC...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/24382274/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5091020-5095073</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72634516-72638569</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>244=6509N871=1X228=1I2707=</td>\n",
       "      <td>2708=1I227=1X871=6509N244=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 4050, 'I': 1, 'X': 1, 'N': 6509}</td>\n",
       "      <td>5091020</td>\n",
       "      <td>72634516</td>\n",
       "      <td>[TGTACATGGGAAAAG, GTACATGGGAAAAGA, TACATGGGAAA...</td>\n",
       "      <td>[TGTACATGGGAAAAG, GTACATGGGAAAAGA, TACATGGGAAA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/71894259/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5091586-5093996</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72636038-72638448</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>1S108=6079N310=1I42=1I48=1I654=1X30=1X14=1I42=...</td>\n",
       "      <td>1S34=1I28=1I441=1I2=1I65=1I37=1I23=1I337=1I76=...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 2, '=': 2391, 'I': 14, 'X': 2, 'N': 6079}</td>\n",
       "      <td>5091585</td>\n",
       "      <td>72636037</td>\n",
       "      <td>[GAGAAGGGGAGGCTC, CGGAAGGGAAAAAAA, GGAAGGGAAAA...</td>\n",
       "      <td>[GAGAAGGGGAGGCTC, CGGAAGGGAAAAAAA, GGAAGGGAAAA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/87031924/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5097771-5102850</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72633735-72638814</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>487S1102=1I649=1D1044=1I1794=</td>\n",
       "      <td>1788=1I1050=1D650=1I1101=487S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 4589, 'I': 2, 'D': 1, 'S': 487}</td>\n",
       "      <td>5097284</td>\n",
       "      <td>72633735</td>\n",
       "      <td>[GGACTGCCCAGCTGG, GACTGCCCAGCTGGG, ACTGCCCAGCT...</td>\n",
       "      <td>[GGACTGCCCAGCTGG, GACTGCCCAGCTGGG, ACTGCCCAGCT...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample                                read          contigHap1  \\\n",
       "46   HG00268  m64119_210712_091919/145165136/ccs  haplotype1-0000001   \n",
       "464  HG00268   m64119_210712_091919/29230379/ccs  haplotype1-0000001   \n",
       "568  HG00268   m64119_210712_091919/64226014/ccs  haplotype1-0000001   \n",
       "569  HG00268  m64119_210712_091919/120588203/ccs  haplotype1-0000001   \n",
       "573  HG00268  m64119_210712_091919/126222455/ccs  haplotype1-0000001   \n",
       "580  HG00268   m64119_210712_091919/26346562/ccs  haplotype1-0000001   \n",
       "588  HG00268   m64119_210712_091919/24382274/ccs  haplotype1-0000001   \n",
       "591  HG00268   m64119_210712_091919/71894259/ccs  haplotype1-0000001   \n",
       "593  HG00268   m64119_210712_091919/87031924/ccs  haplotype1-0000001   \n",
       "\n",
       "        locationHap1          contigHap2       locationHap2  flag1  flag2  \\\n",
       "46     293599-295689  haplotype2-0000064  77437144-77439234     16      0   \n",
       "464  3166242-3171367  haplotype2-0000064  74450787-74455912      0     16   \n",
       "568  5090954-5094461  haplotype2-0000064  72635129-72638636      0     16   \n",
       "569  5090974-5093553  haplotype2-0000064  72636037-72638616      0     16   \n",
       "573  5090979-5095855  haplotype2-0000064  72633734-72638610      0     16   \n",
       "580  5091010-5094148  haplotype2-0000064  72635447-72638585      0     16   \n",
       "588  5091020-5095073  haplotype2-0000064  72634516-72638569      0     16   \n",
       "591  5091586-5093996  haplotype2-0000064  72636038-72638448      0     16   \n",
       "593  5097771-5102850  haplotype2-0000064  72633735-72638814      0     16   \n",
       "\n",
       "                                             hap1CIGAR  \\\n",
       "46                     1S373=2255N1018=2I553=1825N142=   \n",
       "464  479=26221N76=39010N88=48869N349=1I494=1I260=1I...   \n",
       "568                            1S310=6509N1100=1I2094=   \n",
       "569                            1S290=6509N1100=1I1186=   \n",
       "573                        285=6509N7=1D1092=1I3489=1S   \n",
       "580     2S132=1I122=6509N337=1I530=1I233=1I364=1I1412=   \n",
       "588                         244=6509N871=1X228=1I2707=   \n",
       "591  1S108=6079N310=1I42=1I48=1I654=1X30=1X14=1I42=...   \n",
       "593                      487S1102=1I649=1D1044=1I1794=   \n",
       "\n",
       "                                             hap2CIGAR  ...  \\\n",
       "46                 81=1X60=1825N536=1I1036=2243N373=1S  ...   \n",
       "464  436=1I525=1I493=1I1017=1I888=1I266=1I497=1I352...  ...   \n",
       "568                            2095=1I1099=6509N310=1S  ...   \n",
       "569                            1187=1I1099=6509N290=1S  ...   \n",
       "573                        1S3490=1I1090=1D8=6509N285=  ...   \n",
       "580     1409=1I368=1I226=1I531=1I342=6509N118=1I136=2S  ...   \n",
       "588                         2708=1I227=1X871=6509N244=  ...   \n",
       "591  1S34=1I28=1I441=1I2=1I65=1I37=1I23=1I337=1I76=...  ...   \n",
       "593                      1788=1I1050=1D650=1I1101=487S  ...   \n",
       "\n",
       "                                       BasicPhase_H2 Hap1Start Hap2Start  \\\n",
       "46    {'=': 2086, 'X': 1, 'N': 4068, 'I': 1, 'S': 1}    293598  77437144   \n",
       "464         {'=': 5116, 'I': 7, 'N': 114023, 'X': 1}   3166242  74450787   \n",
       "568           {'=': 3504, 'I': 1, 'N': 6509, 'S': 1}   5090953  72635129   \n",
       "569           {'=': 2576, 'I': 1, 'N': 6509, 'S': 1}   5090973  72636037   \n",
       "573   {'S': 1, '=': 4873, 'I': 1, 'D': 1, 'N': 6509}   5090979  72633733   \n",
       "580           {'=': 3130, 'I': 5, 'N': 6509, 'S': 2}   5091008  72635447   \n",
       "588           {'=': 4050, 'I': 1, 'X': 1, 'N': 6509}   5091020  72634516   \n",
       "591  {'S': 2, '=': 2391, 'I': 14, 'X': 2, 'N': 6079}   5091585  72636037   \n",
       "593            {'=': 4589, 'I': 2, 'D': 1, 'S': 487}   5097284  72633735   \n",
       "\n",
       "                                           Hap1Missing  \\\n",
       "46                                   [GACATTTTTAAGTTT]   \n",
       "464  [CATGGCTTCCTCCCC, ATGGCTTCCTCCCCA, TGGCTTCCTCC...   \n",
       "568  [GCCCGCCCGCCCTAG, GGGGAAAAAAAAAAA, GGGAAAAAAAA...   \n",
       "569  [GAGAGCACAGCCGAG, GGGGAAAAAAAAAAA, GGGAAAAAAAA...   \n",
       "573  [GAACCGATTTGTTGT, AACCGATTTGTTGTG, ACCGATTTGTT...   \n",
       "580  [GGGACCCCGGCCAGC, GGACCCCGGCCAGCG, CTCCGCCATTC...   \n",
       "588  [TGTACATGGGAAAAG, GTACATGGGAAAAGA, TACATGGGAAA...   \n",
       "591  [GAGAAGGGGAGGCTC, CGGAAGGGAAAAAAA, GGAAGGGAAAA...   \n",
       "593  [GGACTGCCCAGCTGG, GACTGCCCAGCTGGG, ACTGCCCAGCT...   \n",
       "\n",
       "                                           Hap2Missing  PhaseV1Results  \\\n",
       "46   [GACATTTTTAAGTTT, TCCATGTGGACCGCG, CCATGTGGACC...  NotAbleToPhase   \n",
       "464  [CCAGGAGAGCCCGGA, CAGGAGAGCCCGGAT, AGGAGAGCCCG...  NotAbleToPhase   \n",
       "568  [GCCCGCCCGCCCTAG, TCTTCCCTTGGGGGG, CTTCCCTTGGG...  NotAbleToPhase   \n",
       "569  [GAGAGCACAGCCGAG, TCTTCCCTTGGGGGG, CTTCCCTTGGG...  NotAbleToPhase   \n",
       "573  [GAACCGATTTGTTGT, AACCGATTTGTTGTG, ACCGATTTGTT...  NotAbleToPhase   \n",
       "580  [GGGACCCCGGCCAGC, GGACCCCGGCCAGCG, CTCCGCCATTC...  NotAbleToPhase   \n",
       "588  [TGTACATGGGAAAAG, GTACATGGGAAAAGA, TACATGGGAAA...  NotAbleToPhase   \n",
       "591  [GAGAAGGGGAGGCTC, CGGAAGGGAAAAAAA, GGAAGGGAAAA...  NotAbleToPhase   \n",
       "593  [GGACTGCCCAGCTGG, GACTGCCCAGCTGGG, ACTGCCCAGCT...  NotAbleToPhase   \n",
       "\n",
       "    PhaseV2Results       Breakdown PhaseV1_DeepResults     Final_Phase  \n",
       "46      Haplotype1  noPerfectMatch          Haplotype2  NotAbleToPhase  \n",
       "464     Haplotype1  noPerfectMatch          Haplotype2  NotAbleToPhase  \n",
       "568     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "569     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "573     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "580     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "588     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "591     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "593     Haplotype1  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "\n",
       "[9 rows x 22 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDF[(sampleDF['PhaseV2Results']=='Haplotype1') & (sampleDF['Final_Phase']=='NotAbleToPhase')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "7daea28c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "154\n",
      "159\n"
     ]
    }
   ],
   "source": [
    "for x in sampleDF[(sampleDF['PhaseV2Results']=='Haplotype2') & (sampleDF['Final_Phase']=='NotAbleToPhase')]['Hap2Missing']:\n",
    "    print(len(x))\n",
    "\n",
    "for x in sampleDF[(sampleDF['PhaseV2Results']=='Haplotype2') & (sampleDF['Final_Phase']=='NotAbleToPhase')]['Hap1Missing']:\n",
    "    print(len(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "94de40e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read</th>\n",
       "      <th>contigHap1</th>\n",
       "      <th>locationHap1</th>\n",
       "      <th>contigHap2</th>\n",
       "      <th>locationHap2</th>\n",
       "      <th>flag1</th>\n",
       "      <th>flag2</th>\n",
       "      <th>hap1CIGAR</th>\n",
       "      <th>hap2CIGAR</th>\n",
       "      <th>...</th>\n",
       "      <th>BasicPhase_H2</th>\n",
       "      <th>Hap1Start</th>\n",
       "      <th>Hap2Start</th>\n",
       "      <th>Hap1Missing</th>\n",
       "      <th>Hap2Missing</th>\n",
       "      <th>PhaseV1Results</th>\n",
       "      <th>PhaseV2Results</th>\n",
       "      <th>Breakdown</th>\n",
       "      <th>PhaseV1_DeepResults</th>\n",
       "      <th>Final_Phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/18155590/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5097771-5101120</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>72635111-72638460</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>135S229=1X866=1D2117=</td>\n",
       "      <td>2102=1D881=1X229=135S</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3212, 'D': 1, 'X': 1, 'S': 135}</td>\n",
       "      <td>5097636</td>\n",
       "      <td>72635111</td>\n",
       "      <td>[GGAGACGCGACCAGG, GAGACGCGACCAGGA, AGACGCGACCA...</td>\n",
       "      <td>[GGAGACGCGACCAGG, GAGACGCGACCAGGA, AGACGCGACCA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>noPerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample                               read          contigHap1  \\\n",
       "592  HG00268  m64119_210712_091919/18155590/ccs  haplotype1-0000001   \n",
       "\n",
       "        locationHap1          contigHap2       locationHap2  flag1  flag2  \\\n",
       "592  5097771-5101120  haplotype2-0000064  72635111-72638460      0     16   \n",
       "\n",
       "                 hap1CIGAR              hap2CIGAR  ...  \\\n",
       "592  135S229=1X866=1D2117=  2102=1D881=1X229=135S  ...   \n",
       "\n",
       "                             BasicPhase_H2 Hap1Start Hap2Start  \\\n",
       "592  {'=': 3212, 'D': 1, 'X': 1, 'S': 135}   5097636  72635111   \n",
       "\n",
       "                                           Hap1Missing  \\\n",
       "592  [GGAGACGCGACCAGG, GAGACGCGACCAGGA, AGACGCGACCA...   \n",
       "\n",
       "                                           Hap2Missing  PhaseV1Results  \\\n",
       "592  [GGAGACGCGACCAGG, GAGACGCGACCAGGA, AGACGCGACCA...  NotAbleToPhase   \n",
       "\n",
       "    PhaseV2Results       Breakdown PhaseV1_DeepResults     Final_Phase  \n",
       "592     Haplotype2  noPerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDF[(sampleDF['PhaseV2Results']=='Haplotype2') & (sampleDF['Final_Phase']=='NotAbleToPhase')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "826ea972",
   "metadata": {},
   "outputs": [],
   "source": [
    "tempbreak = []\n",
    "for row in sampleDF.index:\n",
    "    perfectone=0\n",
    "    nonperfectone=0\n",
    "    \n",
    "    perfecttwo=0\n",
    "    nonperfecttwo=0\n",
    "\n",
    "    if len(sampleDF.at[row,'BasicPhase_H1'].keys()) == 1 or  len(sampleDF.at[row,'BasicPhase_H1'].keys()) == 2:\n",
    "        for y in [x for x in sampleDF.at[0,'BasicPhase_H1'].keys()]:\n",
    "            if y == '=' or y == 'N':\n",
    "                perfectone+=1\n",
    "            else:\n",
    "                nonperfectone+=1\n",
    "    else:\n",
    "        nonperfectone+=1\n",
    "     \n",
    "    if len(sampleDF.at[row,'BasicPhase_H2'].keys()) == 1 or  len(sampleDF.at[row,'BasicPhase_H2'].keys()) == 2:\n",
    "        for y in [x for x in sampleDF.at[0,'BasicPhase_H2'].keys()]:\n",
    "            if y == '=' or y == 'N':\n",
    "                perfecttwo+=1\n",
    "            else:\n",
    "                nonperfecttwo+=1\n",
    "    else:\n",
    "        nonperfecttwo+=1\n",
    "        \n",
    "    if (perfectone >0 and nonperfectone==0) and (perfecttwo >0 and nonperfecttwo==0):\n",
    "        tempbreak.append('BothPerfectMatches')\n",
    "    elif (perfectone > 0 and nonperfectone==0 and nonperfecttwo >0) or (perfecttwo > 0  and nonperfecttwo==0 and nonperfectone >0):\n",
    "        tempbreak.append('onePerfectMatch')\n",
    "    else:\n",
    "        tempbreak.append('noPerfectMatch')\n",
    "sampleDF['Breakdown'] =tempbreak "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f0847469",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'noPerfectMatch': 1000})\n"
     ]
    }
   ],
   "source": [
    "print(collections.Counter(sampleDF['Breakdown']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0b888671",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>read</th>\n",
       "      <th>contigHap1</th>\n",
       "      <th>locationHap1</th>\n",
       "      <th>contigHap2</th>\n",
       "      <th>locationHap2</th>\n",
       "      <th>flag1</th>\n",
       "      <th>flag2</th>\n",
       "      <th>hap1CIGAR</th>\n",
       "      <th>hap2CIGAR</th>\n",
       "      <th>...</th>\n",
       "      <th>BasicPhase_H2</th>\n",
       "      <th>Hap1Start</th>\n",
       "      <th>Hap2Start</th>\n",
       "      <th>Hap1Missing</th>\n",
       "      <th>Hap2Missing</th>\n",
       "      <th>PhaseV1Results</th>\n",
       "      <th>PhaseV2Results</th>\n",
       "      <th>Breakdown</th>\n",
       "      <th>PhaseV1_DeepResults</th>\n",
       "      <th>Final_Phase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/132317941/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>116003-118634</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77618176-77620807</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>736=1X1893=</td>\n",
       "      <td>1893=1X736=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2629, 'X': 1}</td>\n",
       "      <td>116003</td>\n",
       "      <td>77618176</td>\n",
       "      <td>[CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...</td>\n",
       "      <td>[CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/126157970/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>117047-120001</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77617939-77620893</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1823=1130S</td>\n",
       "      <td>1130S1823=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'S': 1130, '=': 1823}</td>\n",
       "      <td>117047</td>\n",
       "      <td>77616809</td>\n",
       "      <td>[CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...</td>\n",
       "      <td>[CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/29100612/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>127485-130280</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77606527-77609322</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2190=1D14=1D129=1D461=</td>\n",
       "      <td>459=1D128=1D15=1D2192=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2794, 'D': 3}</td>\n",
       "      <td>127485</td>\n",
       "      <td>77606527</td>\n",
       "      <td>[GTCCCACAGGCTGGT, TCCCACAGGCTGGTG, CCCACAGGCTG...</td>\n",
       "      <td>[GTCCCACAGGCTGGT, TCCCACAGGCTGGTG, CCCACAGGCTG...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/76744772/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>127485-129330</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77607469-77609314</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>356=1D29=2D1132=2D24=2D44=1D40=1D27=1D56=1D136=</td>\n",
       "      <td>135=1D56=1D27=1D39=1D43=2D24=2D1133=2D30=1D357=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 1844, 'D': 11}</td>\n",
       "      <td>127485</td>\n",
       "      <td>77607469</td>\n",
       "      <td>[TTGTCTTTTAAAAGC, TGTCTTTTAAAAGCA, GTCTTTTAAAA...</td>\n",
       "      <td>[TTGTCTTTTAAAAGC, TGTCTTTTAAAAGCA, GTCTTTTAAAA...</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/107544667/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>127485-130948</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>77579955-77583418</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>3205=2419N90=15431N131=8055N36=</td>\n",
       "      <td>36=8055N131=15431N90=2421N3205=</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 3462, 'N': 25907}</td>\n",
       "      <td>127485</td>\n",
       "      <td>77579955</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "      <td>NotAbleToPhase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>915</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/173212438/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779730-5782362</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71930432-71933064</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1123=902N148=309N142=2102N165=2582N161=1026N85...</td>\n",
       "      <td>58=3450N152=1295N144=4431N163=2828N89=1881N201...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2631, 'N': 22071}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71930432</td>\n",
       "      <td>[CTAGTGGAGACTGAG, TAGTGGAGACTGAGG, AGTGGAGACTG...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>916</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/173475725/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779730-5782389</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71930405-71933064</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1123=902N148=309N142=2102N165=2582N161=1026N85...</td>\n",
       "      <td>10=1X74=3450N3=1X3=1X144=1295N144=4431N163=282...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2655, 'X': 3, 'N': 22071}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71930405</td>\n",
       "      <td>[]</td>\n",
       "      <td>[CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/174981513/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779730-5782309</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933935-71936514</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1123=902N148=309N142=2102N165=2582N161=1026N85...</td>\n",
       "      <td>157=1295N144=4431N163=2828N89=1881N201=1434N85...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2578, 'N': 18621}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71933935</td>\n",
       "      <td>[CTAGTGGAGACTGAG, TAGTGGAGACTGAGG, AGTGGAGACTG...</td>\n",
       "      <td>[]</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>Haplotype2</td>\n",
       "      <td>Haplotype2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>925</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/178259518/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779730-5782309</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71933935-71936514</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1123=902N148=309N142=2102N165=2582N161=1026N85...</td>\n",
       "      <td>8=1X3=1X144=1295N144=4431N163=2828N89=1881N201...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2576, 'X': 2, 'N': 18621}</td>\n",
       "      <td>5779730</td>\n",
       "      <td>71933935</td>\n",
       "      <td>[]</td>\n",
       "      <td>[CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>HG00268</td>\n",
       "      <td>m64119_210712_091919/9700049/ccs</td>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>5779731-5782359</td>\n",
       "      <td>haplotype2-0000064</td>\n",
       "      <td>71930435-71933063</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1122=902N148=309N142=2102N165=2582N161=1026N85...</td>\n",
       "      <td>55=3450N3=1X3=1X144=1295N144=4431N163=2828N89=...</td>\n",
       "      <td>...</td>\n",
       "      <td>{'=': 2625, 'N': 22071, 'X': 2}</td>\n",
       "      <td>5779731</td>\n",
       "      <td>71930435</td>\n",
       "      <td>[]</td>\n",
       "      <td>[CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>onePerfectMatch</td>\n",
       "      <td>Haplotype1</td>\n",
       "      <td>Haplotype1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample                                read          contigHap1  \\\n",
       "2    HG00268  m64119_210712_091919/132317941/ccs  haplotype1-0000001   \n",
       "3    HG00268  m64119_210712_091919/126157970/ccs  haplotype1-0000001   \n",
       "8    HG00268   m64119_210712_091919/29100612/ccs  haplotype1-0000001   \n",
       "9    HG00268   m64119_210712_091919/76744772/ccs  haplotype1-0000001   \n",
       "10   HG00268  m64119_210712_091919/107544667/ccs  haplotype1-0000001   \n",
       "..       ...                                 ...                 ...   \n",
       "915  HG00268  m64119_210712_091919/173212438/ccs  haplotype1-0000001   \n",
       "916  HG00268  m64119_210712_091919/173475725/ccs  haplotype1-0000001   \n",
       "920  HG00268  m64119_210712_091919/174981513/ccs  haplotype1-0000001   \n",
       "925  HG00268  m64119_210712_091919/178259518/ccs  haplotype1-0000001   \n",
       "939  HG00268    m64119_210712_091919/9700049/ccs  haplotype1-0000001   \n",
       "\n",
       "        locationHap1          contigHap2       locationHap2  flag1  flag2  \\\n",
       "2      116003-118634  haplotype2-0000064  77618176-77620807     16      0   \n",
       "3      117047-120001  haplotype2-0000064  77617939-77620893     16      0   \n",
       "8      127485-130280  haplotype2-0000064  77606527-77609322     16      0   \n",
       "9      127485-129330  haplotype2-0000064  77607469-77609314     16      0   \n",
       "10     127485-130948  haplotype2-0000064  77579955-77583418     16      0   \n",
       "..               ...                 ...                ...    ...    ...   \n",
       "915  5779730-5782362  haplotype2-0000064  71930432-71933064     16      0   \n",
       "916  5779730-5782389  haplotype2-0000064  71930405-71933064     16      0   \n",
       "920  5779730-5782309  haplotype2-0000064  71933935-71936514     16      0   \n",
       "925  5779730-5782309  haplotype2-0000064  71933935-71936514     16      0   \n",
       "939  5779731-5782359  haplotype2-0000064  71930435-71933063     16      0   \n",
       "\n",
       "                                             hap1CIGAR  \\\n",
       "2                                          736=1X1893=   \n",
       "3                                           1823=1130S   \n",
       "8                               2190=1D14=1D129=1D461=   \n",
       "9      356=1D29=2D1132=2D24=2D44=1D40=1D27=1D56=1D136=   \n",
       "10                     3205=2419N90=15431N131=8055N36=   \n",
       "..                                                 ...   \n",
       "915  1123=902N148=309N142=2102N165=2582N161=1026N85...   \n",
       "916  1123=902N148=309N142=2102N165=2582N161=1026N85...   \n",
       "920  1123=902N148=309N142=2102N165=2582N161=1026N85...   \n",
       "925  1123=902N148=309N142=2102N165=2582N161=1026N85...   \n",
       "939  1122=902N148=309N142=2102N165=2582N161=1026N85...   \n",
       "\n",
       "                                             hap2CIGAR  ...  \\\n",
       "2                                          1893=1X736=  ...   \n",
       "3                                           1130S1823=  ...   \n",
       "8                               459=1D128=1D15=1D2192=  ...   \n",
       "9      135=1D56=1D27=1D39=1D43=2D24=2D1133=2D30=1D357=  ...   \n",
       "10                     36=8055N131=15431N90=2421N3205=  ...   \n",
       "..                                                 ...  ...   \n",
       "915  58=3450N152=1295N144=4431N163=2828N89=1881N201...  ...   \n",
       "916  10=1X74=3450N3=1X3=1X144=1295N144=4431N163=282...  ...   \n",
       "920  157=1295N144=4431N163=2828N89=1881N201=1434N85...  ...   \n",
       "925  8=1X3=1X144=1295N144=4431N163=2828N89=1881N201...  ...   \n",
       "939  55=3450N3=1X3=1X144=1295N144=4431N163=2828N89=...  ...   \n",
       "\n",
       "                       BasicPhase_H2 Hap1Start Hap2Start  \\\n",
       "2                {'=': 2629, 'X': 1}    116003  77618176   \n",
       "3             {'S': 1130, '=': 1823}    117047  77616809   \n",
       "8                {'=': 2794, 'D': 3}    127485  77606527   \n",
       "9               {'=': 1844, 'D': 11}    127485  77607469   \n",
       "10           {'=': 3462, 'N': 25907}    127485  77579955   \n",
       "..                               ...       ...       ...   \n",
       "915          {'=': 2631, 'N': 22071}   5779730  71930432   \n",
       "916  {'=': 2655, 'X': 3, 'N': 22071}   5779730  71930405   \n",
       "920          {'=': 2578, 'N': 18621}   5779730  71933935   \n",
       "925  {'=': 2576, 'X': 2, 'N': 18621}   5779730  71933935   \n",
       "939  {'=': 2625, 'N': 22071, 'X': 2}   5779731  71930435   \n",
       "\n",
       "                                           Hap1Missing  \\\n",
       "2    [CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...   \n",
       "3    [CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...   \n",
       "8    [GTCCCACAGGCTGGT, TCCCACAGGCTGGTG, CCCACAGGCTG...   \n",
       "9    [TTGTCTTTTAAAAGC, TGTCTTTTAAAAGCA, GTCTTTTAAAA...   \n",
       "10                                                  []   \n",
       "..                                                 ...   \n",
       "915  [CTAGTGGAGACTGAG, TAGTGGAGACTGAGG, AGTGGAGACTG...   \n",
       "916                                                 []   \n",
       "920  [CTAGTGGAGACTGAG, TAGTGGAGACTGAGG, AGTGGAGACTG...   \n",
       "925                                                 []   \n",
       "939                                                 []   \n",
       "\n",
       "                                           Hap2Missing  PhaseV1Results  \\\n",
       "2    [CTCTAGAAATATATA, TCTAGAAATATATAT, CTAGAAATATA...  NotAbleToPhase   \n",
       "3    [CTTCCTGTCGGGCTC, TTCCTGTCGGGCTCT, TCCTGTCGGGC...  NotAbleToPhase   \n",
       "8    [GTCCCACAGGCTGGT, TCCCACAGGCTGGTG, CCCACAGGCTG...  NotAbleToPhase   \n",
       "9    [TTGTCTTTTAAAAGC, TGTCTTTTAAAAGCA, GTCTTTTAAAA...  NotAbleToPhase   \n",
       "10                                                  []  NotAbleToPhase   \n",
       "..                                                 ...             ...   \n",
       "915                                                 []      Haplotype2   \n",
       "916  [CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...      Haplotype1   \n",
       "920                                                 []      Haplotype2   \n",
       "925  [CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...      Haplotype1   \n",
       "939  [CTAGTGGAGACTGAC, TAGTGGAGACTGACG, AGTGGAGACTG...      Haplotype1   \n",
       "\n",
       "     PhaseV2Results        Breakdown PhaseV1_DeepResults     Final_Phase  \n",
       "2    NotAbleToPhase  onePerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "3    NotAbleToPhase  onePerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "8    NotAbleToPhase  onePerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "9    NotAbleToPhase  onePerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "10   NotAbleToPhase  onePerfectMatch      NotAbleToPhase  NotAbleToPhase  \n",
       "..              ...              ...                 ...             ...  \n",
       "915      Haplotype2  onePerfectMatch          Haplotype2      Haplotype2  \n",
       "916      Haplotype1  onePerfectMatch          Haplotype1      Haplotype1  \n",
       "920      Haplotype2  onePerfectMatch          Haplotype2      Haplotype2  \n",
       "925      Haplotype1  onePerfectMatch          Haplotype1      Haplotype1  \n",
       "939      Haplotype1  onePerfectMatch          Haplotype1      Haplotype1  \n",
       "\n",
       "[128 rows x 22 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleDF[sampleDF['Breakdown']=='onePerfectMatch']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "633e5c89",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
