{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b1f72f1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'chrM': 37})\n",
      "[('known', 64372), ('auto', 61541), ('notbasic', 46590), ('female', 2669), ('male', 162)]\n"
     ]
    }
   ],
   "source": [
    "import pathlib as pl\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import dnaio\n",
    "import collections as col\n",
    "import io\n",
    "\n",
    "# Source\n",
    "# Gencode V44 basic annotation / gene model\n",
    "# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.basic.annotation.gtf.gz\n",
    "# md5: 7450ef42cf9cb3d29625320b22d4bb45\n",
    "\n",
    "# Gencode v44 protein coding sequences / cDNA\n",
    "# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.pc_transcripts.fa.gz\n",
    "# md5: 1d6f02c471e76f421e36921ddb5715a1\n",
    "\n",
    "# 2023-11-08\n",
    "# Filtered: 37 entries for chrM\n",
    "# Remaining set:\n",
    "# Counter({'known': 64385, 'auto': 61554, 'notbasic': 46577, 'female': 2669, 'male': 162})\n",
    "# (content of object \"count_records\" after run)\n",
    "# leaves at most 64385 - 46577 = 17808 transcripts in output\n",
    "\n",
    "local_wd = pl.Path(\"temp\").resolve()\n",
    "local_gtf = local_wd.joinpath(\"gencode.v44.basic.annotation.gtf.gz\")\n",
    "assert local_gtf.is_file()\n",
    "local_fasta = local_wd.joinpath(\"gencode.v44.pc_transcripts.fa.gz\")\n",
    "assert local_fasta.is_file()\n",
    "\n",
    "def cache_transcript_to_chromosome():\n",
    "\n",
    "    USE_CHROMOSOMES = [f\"chr{i}\" for i in range(1,23)] + [\"chrX\", \"chrY\"]\n",
    "    CHROM_LOC = {\n",
    "        \"chrY\": \"male\",\n",
    "        \"chrX\": \"female\",\n",
    "        \"chrM\": \"mito\"\n",
    "    }\n",
    "\n",
    "    discarded = col.Counter()\n",
    "    transcript_infos = dict()\n",
    "    with gzip.open(local_gtf, \"rt\") as gtf:\n",
    "        for line in gtf:\n",
    "            if line.startswith(\"#\") or not line.strip():\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            if parts[2] != \"transcript\":\n",
    "                continue\n",
    "            if parts[0] not in USE_CHROMOSOMES:\n",
    "                discarded[parts[0]] += 1\n",
    "                continue\n",
    "            position = f\"{parts[0]}:{parts[3]}-{parts[4]}\"\n",
    "            infos = parts[-1].split()\n",
    "            gene_id = None\n",
    "            gene_name = None\n",
    "            gene_type = None\n",
    "            transcript_id = None\n",
    "            for info_type, info_value in zip(infos[:-1], infos[1:]):\n",
    "                if info_type == \"gene_id\":\n",
    "                    gene_id = info_value.strip().strip(';\"')\n",
    "                if info_type == \"gene_name\":\n",
    "                    gene_name = info_value.strip().strip(';\"')\n",
    "                if info_type == \"gene_type\":\n",
    "                    gene_type = info_value.strip().strip(';\"')\n",
    "                if info_type == \"transcript_id\":\n",
    "                    transcript_id = info_value.strip().strip(';\"')\n",
    "                    assert transcript_id not in transcript_infos\n",
    "                if all(x is not None for x in [gene_id, gene_name, gene_type, transcript_id]):\n",
    "                    break\n",
    "            assert all(x is not None for x in [gene_id, gene_name, gene_type, transcript_id])\n",
    "            transcript_infos[transcript_id] = {\n",
    "                \"chrom\": parts[0],\n",
    "                \"position\": position,\n",
    "                \"karyotype\": CHROM_LOC.get(parts[0], \"auto\"),\n",
    "                \"gene_id\": gene_id,\n",
    "                \"gene_name\": gene_name,\n",
    "                \"gene_biotype\": gene_type,\n",
    "                \"transcript_id\": transcript_id\n",
    "            }\n",
    "    print(discarded)\n",
    "    return transcript_infos\n",
    "                    \n",
    "\n",
    "transcript_infos = cache_transcript_to_chromosome()\n",
    "\n",
    "buffer_auto = io.StringIO()\n",
    "buffer_female = io.StringIO()\n",
    "buffer_male = io.StringIO()\n",
    "\n",
    "count_records = col.Counter()\n",
    "record_table = []\n",
    "with dnaio.open(local_fasta) as fasta:\n",
    "    for record in fasta:\n",
    "        transcript_id = record.name.split(\"|\")[0]\n",
    "        try:\n",
    "            transcript_info = transcript_infos[transcript_id]\n",
    "            count_records[\"known\"] += 1\n",
    "        except KeyError:\n",
    "            count_records[\"notbasic\"] += 1\n",
    "            continue\n",
    "        shorter_name = [transcript_info[\"position\"]]\n",
    "        shorter_name.extend([p for p in record.name.strip(\"|\").split(\"|\") if not p.startswith(\"OTT\")])\n",
    "        shorter_name = \"|\".join(shorter_name)\n",
    "        transcript_info[\"fasta_header\"] = shorter_name\n",
    "        \n",
    "        karyotype = transcript_info[\"karyotype\"]\n",
    "        assert karyotype in [\"male\", \"female\", \"auto\"]\n",
    "        if karyotype == \"auto\":\n",
    "            buffer_auto.write(f\">{shorter_name}\\n{record.sequence}\\n\")\n",
    "            count_records[\"auto\"] += 1\n",
    "        if karyotype == \"male\":\n",
    "            buffer_male.write(f\">{shorter_name}\\n{record.sequence}\\n\")\n",
    "            count_records[\"male\"] += 1\n",
    "        if karyotype == \"female\":\n",
    "            buffer_female.write(f\">{shorter_name}\\n{record.sequence}\\n\")\n",
    "            count_records[\"female\"] += 1\n",
    "        record_table.append(transcript_info)\n",
    "\n",
    "record_table = pd.DataFrame.from_records(record_table)\n",
    "\n",
    "out_all = local_wd.joinpath(\"gencode.v44.pc_transcripts.122XY.fa.gz\")\n",
    "with gzip.open(out_all, \"wt\") as dump:\n",
    "    _ = dump.write(buffer_auto.getvalue())\n",
    "    _ = dump.write(buffer_female.getvalue())\n",
    "    _ = dump.write(buffer_male.getvalue())\n",
    "    \n",
    "out_female = local_wd.joinpath(\"gencode.v44.pc_transcripts.122X.fa.gz\")\n",
    "with gzip.open(out_female, \"wt\") as dump:\n",
    "    _ = dump.write(buffer_auto.getvalue())\n",
    "    _ = dump.write(buffer_female.getvalue())\n",
    "\n",
    "out_male = local_wd.joinpath(\"gencode.v44.pc_transcripts.122Y.fa.gz\")\n",
    "with gzip.open(out_male, \"wt\") as dump:\n",
    "    _ = dump.write(buffer_auto.getvalue())\n",
    "    _ = dump.write(buffer_male.getvalue())\n",
    "    \n",
    "out_table = local_wd.joinpath(\"gencode.v44.pc_transcripts.tsv.gz\")\n",
    "record_table.to_csv(out_table, sep=\"\\t\", header=True, index=False)\n",
    "\n",
    "print(count_records.most_common())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
