{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db925000",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "No active exception to reraise",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 53\u001b[0m\n\u001b[1;32m     50\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28msorted\u001b[39m(misses\u001b[38;5;241m.\u001b[39mkeys()):\n\u001b[1;32m     51\u001b[0m         dump\u001b[38;5;241m.\u001b[39mwrite(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mmisses[name]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 53\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[1;32m     54\u001b[0m boxes_by_chrom \u001b[38;5;241m=\u001b[39m col\u001b[38;5;241m.\u001b[39mdefaultdict(\u001b[38;5;28mlist\u001b[39m)\n\u001b[1;32m     55\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene_name, missed_haps \u001b[38;5;129;01min\u001b[39;00m misses\u001b[38;5;241m.\u001b[39mmost_common():\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No active exception to reraise"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pathlib as pl\n",
    "import collections as col\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "input_folder = pl.Path(\"/home/ebertp/work/projects/hgsvc/2023_assm_gene/asmgene_raw\")\n",
    "\n",
    "dups = col.Counter()\n",
    "dups_iv = dict()\n",
    "misses = col.Counter()\n",
    "misses_iv = dict()\n",
    "total_haps = 0\n",
    "for txt_file in input_folder.glob(\"*asm-hap*.txt\"):\n",
    "    \n",
    "    with open(txt_file, \"r\") as listing:\n",
    "        total_haps += 1\n",
    "        for line in listing:\n",
    "            parts = line.strip().split()\n",
    "            if parts[0] in [\"0\", \"M\", \"D\"]:\n",
    "                \n",
    "                bucket = dups if parts[0] == \"D\" else misses\n",
    "                regions = dups_iv if parts[0] == \"D\" else misses_iv\n",
    "                \n",
    "                id_parts = parts[2].strip(\"|\").split(\"|\")\n",
    "                biotype = id_parts[-1]\n",
    "                length = int(id_parts[-2])\n",
    "                gene_name = id_parts[-3]\n",
    "                if gene_name.startswith(\"MT\"):\n",
    "                    continue\n",
    "                if parts[-3] in [\"chrX\", \"chrY\", \"chrM\", \"chrMT\"]:\n",
    "                    continue\n",
    "                region = parts[-3], int(parts[-2]), int(parts[-1]), gene_name, length\n",
    "                if \"protein_coding\" in biotype:\n",
    "                    bucket[gene_name] += 1\n",
    "                    regions[gene_name] = region\n",
    "                elif (\"TR_\" in biotype or \"IG_\" in biotype) and \"pseudogene\" not in biotype:\n",
    "                    bucket[gene_name] += 1\n",
    "                    regions[gene_name] = region\n",
    "                else:\n",
    "                    bucket[\"other\"] += 1                \n",
    "            else:\n",
    "                continue\n",
    "\n",
    "bar_heights = []\n",
    "bar_labels = []\n",
    "\n",
    "with open(\"missing_genes.tsv\", \"w\") as dump:\n",
    "    for name in sorted(misses.keys()):\n",
    "        dump.write(f\"{name}\\t{misses[name]}\\n\")\n",
    "    \n",
    "raise\n",
    "boxes_by_chrom = col.defaultdict(list)\n",
    "for gene_name, missed_haps in misses.most_common():\n",
    "    if gene_name == \"other\":\n",
    "        continue\n",
    "    pct_missed_haps = round(missed_haps / total_haps * 100, 2)\n",
    "    region = misses_iv[gene_name]\n",
    "    boxes_by_chrom[region[0]].append(pct_missed_haps)\n",
    "    \n",
    "    bar_heights.append(pct_missed_haps)\n",
    "    bar_labels.append(gene_name)\n",
    "    \n",
    "fig, ax = plt.subplots(figsize=(12,8))\n",
    "\n",
    "_ = ax.bar(\n",
    "    list(range(1, len(bar_heights)+1)),\n",
    "    bar_heights,\n",
    "    align=\"center\",\n",
    "    width=1\n",
    ")\n",
    "\n",
    "ax.set_xlabel(\"Distinct coding genes (unlabeled)\", fontsize=14)\n",
    "ax.set_ylabel(\"Missing in haplotype assembly (%)\", fontsize=14)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12,8))\n",
    "\n",
    "_ = ax.bar(\n",
    "    list(range(1, 51)),\n",
    "    bar_heights[:50],\n",
    "    align=\"center\",\n",
    "    width=1\n",
    ")\n",
    "\n",
    "ax.set_xticks(list(range(1, 51)))\n",
    "ax.set_xticklabels(bar_labels[:50], rotation=90, fontsize=12)\n",
    "\n",
    "ax.set_xlabel(\"Distinct coding genes (first 50)\", fontsize=14)\n",
    "ax.set_ylabel(\"Missing in haplotype assembly (%)\", fontsize=14)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12,8))\n",
    "\n",
    "boxes = [[]] * 22\n",
    "box_labels = [[]] * 22\n",
    "for chrom, values in boxes_by_chrom.items():\n",
    "    idx = int(chrom.strip(\"chr\")) - 1\n",
    "    boxes[idx] = values\n",
    "    \n",
    "    pct_of_genes = round(len(values) / len(bar_heights) * 100, 1)\n",
    "    box_label = f\"{pct_of_genes}% - {chrom.strip('chr')}\"\n",
    "    box_labels[idx] = box_label\n",
    "\n",
    "_ = ax.boxplot(\n",
    "    boxes,\n",
    "    positions=list(range(1, 23)),\n",
    ")\n",
    "\n",
    "ax.set_xticks(list(range(1, 23)))\n",
    "ax.set_xticklabels(box_labels, rotation=90)\n",
    "\n",
    "ax.set_xlabel(\"Missing genes per chromosome (% of all distinct missing)\", fontsize=14)\n",
    "ax.set_ylabel(\"Missing in haplotype assembly (%)\", fontsize=14)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
