{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0838c1b6-026c-41b3-9cf3-3a3ad9ee5b0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "verkko median aun: 137.0 Mbp\n",
      "hifiasm median aun: 95.0 Mbp\n",
      "verkko median variant QV: 54.0\n",
      "verkko median Merqury QV: 57.0\n",
      "hifiasm median variant QV: 53.0\n",
      "hifiasm median Merqury QV: 58.0\n",
      "verkko median Hap: 99.7%\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pathlib\n",
    "import math\n",
    "import numpy as np\n",
    "import collections as col\n",
    "import pickle as pck\n",
    "\n",
    "%cd -q \"/home/ebertp/work/code/cubi/project-run-hgsvc-hybrid-assemblies/notebooks\"\n",
    "\n",
    "_PROJECT_CONFIG_NB = str(pathlib.Path(\"00_project_config.ipynb\").resolve(strict=True))\n",
    "_ASSM_STATS_NB = str(pathlib.Path(\"10_assm_stats.ipynb\").resolve(strict=True))\n",
    "\n",
    "%run $_PROJECT_CONFIG_NB\n",
    "%run $_ASSM_STATS_NB\n",
    "\n",
    "def get_assm_stat(assembler, statistic):\n",
    "\n",
    "    if assembler == \"verkko\":\n",
    "        table = VRK_ASSM_STATS\n",
    "    if assembler == \"hifiasm\":\n",
    "        table = HSM_ASSM_STATS\n",
    "    statistics = []\n",
    "    for sample in HGSVC_FEMALES + HGSVC_MALES:\n",
    "        for hap in [\"hap1\", \"hap2\"]:\n",
    "            aun = table.get_stat(sample, (hap, statistic), int(1e6))\n",
    "            statistics.append(aun)\n",
    "    median = np.median(np.array(statistics, dtype=int))\n",
    "    print(f\"{assembler} median {statistic}: {median} Mbp\") \n",
    "    return\n",
    "\n",
    "get_assm_stat(\"verkko\", \"aun\")\n",
    "get_assm_stat(\"hifiasm\", \"aun\")\n",
    "\n",
    "def get_qv_stat(assembler):\n",
    "    table_file = PROJECT_BASE.joinpath(\n",
    "        \"annotations\", \"autogen\", f\"{assembler}_qv-est.tsv\"\n",
    "    )\n",
    "    table = pd.read_csv(table_file, sep=\"\\t\", header=0, comment=\"#\")\n",
    "    sample_suffix = None\n",
    "    if assembler == \"verkko\":\n",
    "        sample_suffix = \"vrk-ps-sseq\"\n",
    "    if assembler == \"hifiasm\":\n",
    "        sample_suffix = \"hsm=-vrk-sseq\"\n",
    "    phased_units = table.loc[table[\"asm_unit\"] == \"phased\", :].copy()\n",
    "    assert phased_units.shape[0] == HGSVC_TOTAL\n",
    "    var_qv = round(phased_units[\"variant_qv_est\"].median(), 0)\n",
    "    merq_qv = round(phased_units[\"merqury_qv_est\"].median(), 0)\n",
    "    print(f\"{assembler} median variant QV: {var_qv}\")\n",
    "    print(f\"{assembler} median Merqury QV: {merq_qv}\")\n",
    "    return\n",
    "\n",
    "get_qv_stat(\"verkko\")\n",
    "get_qv_stat(\"hifiasm\")\n",
    "\n",
    "def compute_flagger_stats(assembler):\n",
    "\n",
    "    def assign_au(contig):\n",
    "        if any(contig.startswith(x) for x in [\"h1tg\", \"haplotype1\"]):\n",
    "            return \"hap1\"\n",
    "        elif any(contig.startswith(x) for x in [\"h2tg\", \"haplotype2\"]):\n",
    "            return \"hap2\"\n",
    "        elif \"unassigned\" in contig:\n",
    "            return \"unassigned\"\n",
    "        else:\n",
    "            raise\n",
    "\n",
    "    def ann_seq_length(row):\n",
    "        l = SEQLENS[(row.assembly, None, row.seq)]\n",
    "        return l\n",
    "\n",
    "    CACHE_SEQLEN_FILE = PROJECT_NB_CACHE.joinpath(f\"cache.seqlens.{assembler}.pck\")\n",
    "\n",
    "    SEQLENS = pck.load(open(CACHE_SEQLEN_FILE, \"rb\"))[\"data\"]\n",
    "\n",
    "    folder = pathlib.Path(\n",
    "        f\"/home/ebertp/work/share/hgsvc3/assembly_annotations/flagger/{assembler}\"\n",
    "    )\n",
    "    if assembler == \"verkko\":\n",
    "        assm_suffix = \".vrk-ps-sseq\"\n",
    "    if assembler == \"hifiasm\":\n",
    "        assm_suffix = \".hsm-ps-sseq\"\n",
    "\n",
    "    full_output = []\n",
    "    agg_lengths = dict()\n",
    "    agg_pct = col.defaultdict(list)\n",
    "    for bed_file in folder.glob(\"*.bed\"):\n",
    "        df = pd.read_csv(bed_file, sep=\"\\t\", skiprows=1, header=None, usecols=[0,1,2,3])\n",
    "        sample = bed_file.name.split(\".\")[0]\n",
    "        df.columns = [\"seq\", \"start\", \"end\", \"label\"]\n",
    "        df[\"sample\"] = sample\n",
    "        df[\"asm_unit\"] = df[\"seq\"].apply(assign_au)\n",
    "        df[\"assembly\"] = df[\"sample\"] + assm_suffix\n",
    "        select_large = df[\"end\"] > int(1e6)-1\n",
    "        df[\"is_large\"] = 0\n",
    "        df.loc[select_large, \"is_large\"] = 1\n",
    "        df[\"length\"] = df[\"end\"] - df[\"start\"]\n",
    "        df[\"seq_length\"] = df.apply(ann_seq_length, axis=1)\n",
    "        for au, seqs in df.groupby(\"asm_unit\"):\n",
    "            sub = seqs.drop_duplicates(\"seq\", inplace=False)\n",
    "            agg_len = sub[\"seq_length\"].sum()\n",
    "            agg_lengths[(sample, au)] = int(agg_len)\n",
    "        full_output.append(df)\n",
    "    full_output = pd.concat(full_output, axis=0, ignore_index=False)\n",
    "    \n",
    "    by_au = full_output.groupby([\"sample\", \"asm_unit\", \"label\"])[\"length\"].sum()\n",
    "    for (sample, au, label), value in by_au.items():\n",
    "        sample_au_len = agg_lengths[(sample, au)]\n",
    "        pct_len = round(value/sample_au_len * 100, 2)\n",
    "        agg_pct[(au, label)].append(pct_len)\n",
    "        if au != \"unassigned\":\n",
    "            agg_pct[(\"phased\", label)].append(pct_len)\n",
    "\n",
    "    for label in [\"Hap\"]:\n",
    "        median_pct = round(np.median(agg_pct[(\"phased\", label)]), 1)\n",
    "        print(f\"{assembler} median {label}: {median_pct}%\")\n",
    "\n",
    "    return\n",
    "\n",
    "compute_flagger_stats(\"verkko\")\n",
    "#compute_flagger_stats(\"hifiasm\")\n",
    "\n",
    "def compute_mutual_support(assembler):\n",
    "\n",
    "    summary_files = [\n",
    "        pl.Path(\n",
    "            \"/home/ebertp/work/projects/hgsvc/2024_asm_compare/stats/hsm-to-vrk/summary\",\n",
    "            \"SAMPLES.hsm-to-vrk.mapq-1.seq-100k.aln-10k.merged-regions.tsv\"\n",
    "        ),\n",
    "        pl.Path(\n",
    "            \"/home/ebertp/work/projects/hgsvc/2024_asm_compare/stats/vrk-to-hsm/summary\",\n",
    "            \"SAMPLES.vrk-to-hsm.mapq-1.seq-100k.aln-10k.merged-regions.tsv\"\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    if assembler == \"verkko\":\n",
    "        summary_files = summary_files[:1]\n",
    "    if assembler == \"hifiasm\":\n",
    "        summary_files = summary_files[1:]\n",
    "\n",
    "    for summary_file in summary_files:\n",
    "        df = pd.read_csv(summary_file, sep=\"\\t\", header=0)\n",
    "        qry_trg = summary_file.name.split(\".\")[1]\n",
    "        print(qry_trg)\n",
    "    \n",
    "        for prec, stats in df.groupby(\"precision\"):\n",
    "            print(prec)\n",
    "            print(stats[\"supported_pct\"].median(), \"%\")\n",
    "        print(\"===\")\n",
    "        \n",
    "#compute_mutual_support(\"verkko\")\n",
    "#compute_mutual_support(\"hifiasm\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
