{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d1310fec-5ad5-48f1-a094-f05e8211c0c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "median QV variant:  53.37269860193465\n",
      "median QV k-mer:  58.46979496878602\n"
     ]
    }
   ],
   "source": [
    "import pathlib\n",
    "import functools\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patch\n",
    "import math\n",
    "import random\n",
    "import numpy as np\n",
    "import scipy.stats as stat\n",
    "\n",
    "random.seed()\n",
    "\n",
    "%cd -q \"/home/ebertp/work/code/cubi/project-run-hgsvc-hybrid-assemblies/notebooks\"\n",
    "_PROJECT_CONFIG_NB = str(pathlib.Path(\"00_project_config.ipynb\").resolve(strict=True))\n",
    "_PLOT_CONFIG_NB = str(pathlib.Path(\"05_plot_config.ipynb\").resolve(strict=True))\n",
    "_ASSM_STATS_NB = str(pathlib.Path(\"10_assm_stats.ipynb\").resolve(strict=True))\n",
    "\n",
    "%run $_PROJECT_CONFIG_NB\n",
    "%run $_PLOT_CONFIG_NB\n",
    "%run $_ASSM_STATS_NB\n",
    "\n",
    "_MYNAME=\"plot-qv-estimates\"\n",
    "_MYSTAMP=get_nb_stamp(_MYNAME)\n",
    "\n",
    "_MY_OUT_PATH = PLOT_OUT_MAIN_FIG1.joinpath(\"panels\")\n",
    "\n",
    "ASSEMBLER = \"verkko\"\n",
    "assert ASSEMBLER in [\"verkko\", \"hifiasm\"]\n",
    "\n",
    "\n",
    "def compute_qv(num_errors, ref_size):\n",
    "    p = num_errors / ref_size\n",
    "    try:\n",
    "        q = -10 * math.log10(p)\n",
    "    except ValueError:\n",
    "        return 99\n",
    "    return int(round(q, 0))\n",
    "\n",
    "\n",
    "def deselect_old_version(sample):\n",
    "\n",
    "    # 2024-12-06 fix\n",
    "    # the Verkko Merqury summary tables contain the QV data\n",
    "    # for both HG00514/v1 and HG00514/v2 (fix), need to drop\n",
    "    # the former and rename the latter\n",
    "\n",
    "    if \"HG00514-fix\" in sample:\n",
    "        return True\n",
    "    elif \"HG00514\" in sample:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "\n",
    "def read_merqury_qv_estimates(assembler=ASSEMBLER, aggregate=\"sample\"):\n",
    "\n",
    "    table_files = PROJECT_DATA_ROOT.joinpath(\n",
    "            \"2024_merqury\"\n",
    "    ).glob(f\"{assembler}*.tab\")\n",
    "\n",
    "    summary_columns = [\n",
    "        \"sample\", \"asm_unit\", \"merqury_error_kmer\", \"merqury_total_kmer\",\n",
    "        \"merqury_qv_est\", \"merqury_error_rate\"\n",
    "    ]\n",
    "    contig_columns = [\n",
    "        \"sample\", \"sequence\", \"merqury_error_kmer\", \"merqury_total_kmer\",\n",
    "        \"merqury_qv_est\", \"merqury_error_rate\"\n",
    "    ]\n",
    "    \n",
    "    concat = []\n",
    "    for table_file in table_files:\n",
    "        if \"contigs\" in table_file.name:\n",
    "            use_names = contig_columns\n",
    "        else:\n",
    "            use_names = summary_columns\n",
    "        df = pd.read_csv(\n",
    "            table_file, sep=\"\\t\",\n",
    "            header=None, names=use_names\n",
    "        )\n",
    "        if \"asm_unit\" not in df.columns:\n",
    "            df[\"asm_unit\"] = df[\"sequence\"].apply(set_assembly_unit)\n",
    "        df[\"assembler\"] = assembler\n",
    "        no_errors = df[\"merqury_error_kmer\"] == 0\n",
    "        df.loc[no_errors, \"merqury_qv_est\"] = 99\n",
    "        df[\"merqury_qv_est\"] = df[\"merqury_qv_est\"].astype(float)\n",
    "        if assembler == \"verkko\":\n",
    "            df[\"sample\"] = df[\"sample\"].apply(lambda s: s.replace(\"-unassigned\", \"\"))\n",
    "        concat.append(df)      \n",
    "    \n",
    "    mrg = pd.concat(concat, axis=0, ignore_index=False)\n",
    "    no_seq = pd.isnull(mrg[\"sequence\"])\n",
    "    mrg.loc[no_seq, \"sequence\"] = mrg.loc[no_seq, \"asm_unit\"]\n",
    "\n",
    "    # 2024-12-06 fix entries / keep only Verkko HG00514 v2\n",
    "    if assembler == \"verkko\":\n",
    "        drop_old_hg00514 = mrg[\"sample\"].apply(deselect_old_version)\n",
    "        mrg = mrg.loc[drop_old_hg00514, :].copy()\n",
    "        mrg[\"sample\"] = mrg[\"sample\"].replace({\"HG00514-fix\": \"HG00514\"})\n",
    "    \n",
    "    add_phased = []\n",
    "    for sample, qv_est in mrg.groupby(\"sample\"):\n",
    "        select_h1 = qv_est[\"sequence\"] == \"hap1\"\n",
    "        select_h2 = qv_est[\"sequence\"] == \"hap2\"\n",
    "        \n",
    "        qv1 = qv_est.loc[select_h1, \"merqury_qv_est\"].iloc[0]\n",
    "        qv2 = qv_est.loc[select_h2, \"merqury_qv_est\"].iloc[0]\n",
    "        wt_qv = round((qv1 + qv2)/2, 2)\n",
    "        \n",
    "        add_phased.append(\n",
    "            (sample, \"phased\", \"total\", -1, -1, wt_qv, -1, assembler)\n",
    "        )\n",
    "    add_phased = pd.DataFrame.from_records(add_phased,\n",
    "        columns=[\n",
    "            \"sample\", \"asm_unit\", \"sequence\",\n",
    "            \"merqury_error_kmer\", \"merqury_total_kmer\",\n",
    "            \"merqury_qv_est\", \"merqury_error_rate\", \"assembler\"\n",
    "        ]\n",
    "    )\n",
    "    mrg = pd.concat([mrg, add_phased], axis=0, ignore_index=False)\n",
    "    mrg.sort_values([\"sample\", \"asm_unit\", \"sequence\"], inplace=True)\n",
    "        \n",
    "    if aggregate == \"sample\":\n",
    "        mrg = mrg.loc[mrg[\"asm_unit\"] == \"phased\", :].copy()\n",
    "        assert mrg.shape[0] == HGSVC_TOTAL\n",
    "    \n",
    "    #df[\"qv_est\"] = df[\"qv_est\"].round(0).astype(int)\n",
    "    #df.rename({\"qv_est\": \"merqury_qv_est\"}, axis=1, inplace=True)\n",
    "    return mrg\n",
    "\n",
    "\n",
    "def set_assembly_unit(seq_name):\n",
    "\n",
    "    if any(h in seq_name for h in [\"h1tg\", \"haplotype1\"]):\n",
    "        return \"hap1\"\n",
    "    elif any(h in seq_name for h in [\"h2tg\", \"haplotype2\"]):\n",
    "        return \"hap2\"\n",
    "    elif \"unassigned\" in seq_name:\n",
    "        return \"unassigned\"\n",
    "    elif \"genome\" in seq_name:\n",
    "        return \"wg\"\n",
    "    else:\n",
    "        raise ValueError(seq_name)\n",
    "\n",
    "\n",
    "def read_variant_qv_estimates(assembler=ASSEMBLER, aggregate=\"sample\"):\n",
    "    \"\"\"For Verkko, add QV estimates (as weighted averages)\n",
    "    for the respective assembly units (hap1, hap2, phased etc.)\n",
    "    in the same way as for Merqury\n",
    "    \"\"\"\n",
    "\n",
    "    root_folder = PROJECT_DATA_ROOT.joinpath(\n",
    "        \"2023_var_qv_est\", \"hgsvc\", f\"{assembler}\"\n",
    "    )\n",
    "    table_files = list(root_folder.glob(\"*.tsv\"))\n",
    "    assert len(table_files) == HGSVC_TOTAL\n",
    "    merge = []\n",
    "    combinations = [\n",
    "        (\"hap1\",), (\"hap2\",), (\"hap1\", \"hap2\"),\n",
    "        (\"unassigned\",)\n",
    "    ]\n",
    "    labels = [\"hap1\", \"hap2\", \"phased\", \"unassigned\"]\n",
    "    for table_file in table_files:\n",
    "        sample = table_file.name.rsplit(\".\", 4)[0]\n",
    "        assembly = table_file.name.rsplit(\".\", 3)[0]\n",
    "        df = pd.read_csv(table_file, sep=\"\\t\", header=0)\n",
    "        df[\"asm_unit\"] = df[\"seq_name\"].apply(set_assembly_unit)\n",
    "        df[\"seq_name\"] = df[\"seq_name\"].replace({\"genome\": \"total\"}, inplace=False)\n",
    "        # in the Merqury post-processing, the global QV estimates\n",
    "        # are simply derived as weighted averages; implement the\n",
    "        # same strategy here for comparability\n",
    "        global_estimates = []\n",
    "        for comb, label in zip(combinations, labels):\n",
    "            subset = df.loc[df[\"asm_unit\"].isin(comb), :]\n",
    "            if subset.empty:\n",
    "                continue\n",
    "            total_error = subset[\"num_errors\"].sum()\n",
    "            total_adj_len = subset[\"adj_length\"].sum()\n",
    "            total_seq_len = subset[\"seq_length\"].sum()\n",
    "            qv_est = round(np.average(subset[\"qv\"].values, weights=subset[\"adj_length\"].values), 4)\n",
    "            global_estimates.append(\n",
    "                [sample, label, \"total\", total_seq_len, total_adj_len, total_error, qv_est]\n",
    "            )\n",
    "        global_estimates = pd.DataFrame(\n",
    "            global_estimates,\n",
    "            columns=[\"sample\", \"asm_unit\", \"seq_name\", \"seq_length\", \"adj_length\", \"num_errors\", \"qv\"]\n",
    "        )\n",
    "        df[\"sample\"] = sample\n",
    "        df[\"assembly\"] = assembly\n",
    "        df = pd.concat([df, global_estimates], axis=0, ignore_index=False)\n",
    "        df.rename({\"seq_name\": \"sequence\"}, axis=1, inplace=True)\n",
    "        merge.append(df)\n",
    "    merge = pd.concat(merge, axis=0, ignore_index=False)\n",
    "    merge.rename({\"qv\": \"variant_qv_est\"}, axis=1, inplace=True)\n",
    "    if aggregate == \"sample\":\n",
    "        merge = merge.loc[merge[\"asm_unit\"] == \"phased\", :].copy()\n",
    "        assert merge.shape[0] == HGSVC_TOTAL\n",
    "    return merge        \n",
    "\n",
    "\n",
    "def cache_qv_estimate_table(assembler=ASSEMBLER, aggregate=\"sample\"):\n",
    "\n",
    "    if aggregate == \"sample\":\n",
    "        data_file = PROJECT_NB_CACHE.joinpath(f\"cache.{assembler}-qvest.tsv.gz\")\n",
    "        if not data_file.is_file():\n",
    "            merq_qv = read_merqury_qv_estimates(aggregate=\"sample\")\n",
    "            var_qv = read_variant_qv_estimates(aggregate=\"sample\")\n",
    "            qv_est = merq_qv.merge(var_qv, left_on=[\"sample\", \"sequence\", \"asm_unit\"], right_on=[\"sample\", \"sequence\", \"asm_unit\"])\n",
    "            qv_est.to_csv(data_file, sep=\"\\t\", header=True, index=False)\n",
    "    elif aggregate == \"sequence\":\n",
    "        # create full table output / supplementary table\n",
    "        data_file = PROJECT_BASE.joinpath(\n",
    "            \"annotations\", \"autogen\", f\"{ASSEMBLER}_qv-est.tsv\"\n",
    "        )\n",
    "        if not data_file.is_file():\n",
    "            merq_qv = read_merqury_qv_estimates(aggregate=\"sequence\")\n",
    "            var_qv = read_variant_qv_estimates(aggregate=\"sequence\")\n",
    "            qv_est = merq_qv.merge(var_qv, left_on=[\"sample\", \"sequence\", \"asm_unit\"], right_on=[\"sample\", \"sequence\", \"asm_unit\"])\n",
    "            qv_est.rename(\n",
    "                {\n",
    "                    \"num_errors\": \"variant_num_errors\",\n",
    "                    \"error_bp\": \"merqury_error_bp\",\n",
    "                    \"error_rate\": \"merqury_error_rate\",\n",
    "                    \"total_adj_bp\": \"merqury_adj_length\",\n",
    "                    \"adj_length\": \"variant_adj_length\"\n",
    "                }, axis=1, inplace=True\n",
    "            )\n",
    "            with open(data_file, \"w\") as table:\n",
    "                _ = table.write(\"# AUTOGEN TABLE - DO NOT EDIT\\n\")\n",
    "                _ = table.write(f\"# {_MYSTAMP}\\n\")\n",
    "                qv_est.to_csv(table, sep=\"\\t\", header=True, index=False)\n",
    "    else:\n",
    "        raise ValueError(aggregate)\n",
    "    \n",
    "    return data_file\n",
    "       \n",
    "\n",
    "def add_vector_jitter(values):\n",
    "    return [v+random.gauss(0, 0.2) for v in values]\n",
    "\n",
    "\n",
    "def plot_sample_vs_sample_qv(axes):\n",
    "\n",
    "    qv_cache_file = cache_qv_estimate_table(aggregate=\"sample\")\n",
    "    qv_data = pd.read_csv(qv_cache_file, sep=\"\\t\", header=0)\n",
    "    qv_data[\"sid\"] = qv_data[\"sample\"].apply(lambda x: x.split(\".\")[0])\n",
    "    qv_data.set_index(\"sid\", inplace=True)\n",
    "    \n",
    "    female_qv_var = add_vector_jitter([qv_data.at[sample, \"variant_qv_est\"] for sample in HGSVC_FEMALES])\n",
    "    female_qv_kmer = add_vector_jitter([qv_data.at[sample, \"merqury_qv_est\"] for sample in HGSVC_FEMALES])\n",
    "    #female_colors = [get_super_color(sample) for sample in HGSVC_FEMALES]\n",
    "    female_colors = [get_pop_color(sample) for sample in HGSVC_FEMALES]\n",
    "\n",
    "    # 2024-05 --- for some reason, HG00514 is a strong outlier; make this explicit\n",
    "    # and adjust axes limits ignoring that sample\n",
    "    hg00514_x = int(round(qv_data.at[\"HG00514\", \"variant_qv_est\"], 0))\n",
    "    hg00514_y = int(round(qv_data.at[\"HG00514\", \"merqury_qv_est\"], 0))\n",
    "    \n",
    "    median_female_var = sorted(female_qv_var)[HGSVC_FEMALE//2]\n",
    "    median_female_kmer = sorted(female_qv_kmer)[HGSVC_FEMALE//2]\n",
    "    \n",
    "    male_qv_var = add_vector_jitter([qv_data.at[sample, \"variant_qv_est\"] for sample in HGSVC_MALES])\n",
    "    male_qv_kmer = add_vector_jitter([qv_data.at[sample, \"merqury_qv_est\"] for sample in HGSVC_MALES])\n",
    "    #male_colors = [get_super_color(sample) for sample in HGSVC_MALES]\n",
    "    male_colors = [get_pop_color(sample) for sample in HGSVC_MALES]\n",
    "    \n",
    "    median_male_var = sorted(male_qv_var)[HGSVC_MALE//2]\n",
    "    median_male_kmer = sorted(male_qv_kmer)[HGSVC_MALE//2]\n",
    "\n",
    "    median_all_x = sorted(male_qv_var + female_qv_var)[HGSVC_TOTAL//2]\n",
    "    median_all_y = sorted(male_qv_kmer + female_qv_kmer)[HGSVC_TOTAL//2]\n",
    "\n",
    "    print(\"median QV variant: \", median_all_x)\n",
    "    print(\"median QV k-mer: \", median_all_y)\n",
    "    \n",
    "    axes.scatter(\n",
    "        female_qv_var,\n",
    "        female_qv_kmer,\n",
    "        c=female_colors,\n",
    "        label=\"female\",\n",
    "        marker=FEMALE_MARKER,\n",
    "        edgecolors=\"none\"\n",
    "    )\n",
    "    \n",
    "    axes.scatter(\n",
    "        male_qv_var,\n",
    "        male_qv_kmer,\n",
    "        c=male_colors,\n",
    "        label=\"male\",\n",
    "        marker=MALE_MARKER,\n",
    "        edgecolors=\"none\"\n",
    "    )\n",
    "\n",
    "    # add arrow for HG00514\n",
    "    #axes.annotate(\n",
    "    #    f\"Outlier not shown\\nHG00514 x:{hg00514_x} / y:{hg00514_y}\",\n",
    "    #    (50.5, 52.5), fontsize=10\n",
    "    #)\n",
    "    \n",
    "    #ax.axhline(median_fy, 0, 0.95, zorder=0, color=\"grey\", ls=\"dashed\")\n",
    "    #ax.axvline(median_fx, 0, 0.95, zorder=0, color=\"grey\", ls=\"dashed\")\n",
    "    axes.axhline(median_all_y, 0.01, 0.99, zorder=0, color=\"grey\", ls=\"dashed\")\n",
    "    axes.axvline(median_all_x, 0.01, 0.99, zorder=0, color=\"grey\", ls=\"dashed\")\n",
    "        \n",
    "    axes.set_xlabel(\"QV estimate (variant-based)\")\n",
    "    axes.set_ylabel(\"QV estimate (kmer-based)\")\n",
    "    \n",
    "    axes.spines[\"top\"].set_visible(False)\n",
    "    axes.spines[\"right\"].set_visible(False)\n",
    "    \n",
    "    handles = get_line_legend(\n",
    "        [\n",
    "            {\n",
    "                \"marker\": FEMALE_MARKER,\n",
    "                \"label\": \"female\",\n",
    "                \"linestyle\": \"\",\n",
    "                \"color\": \"black\"\n",
    "            },\n",
    "            {\n",
    "                \"marker\": MALE_MARKER,\n",
    "                \"label\": \"male\",\n",
    "                \"linestyle\": \"\",\n",
    "                \"color\": \"black\"\n",
    "            },\n",
    "            {\n",
    "                \"linestyle\": \"dashed\",\n",
    "                \"label\": \"median\",\n",
    "                \"color\": \"grey\"\n",
    "            }\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    axes.legend(handles=handles, loc=\"lower right\")\n",
    "\n",
    "    if False:\n",
    "        all_x_vals = male_qv_var + female_qv_var\n",
    "        min_x = min(all_x_vals)\n",
    "        max_x = max(all_x_vals)\n",
    "        print(min_x)\n",
    "        print(max_x)\n",
    "        \n",
    "        all_y_vals = male_qv_kmer + female_qv_kmer\n",
    "        min_y = min(all_y_vals)\n",
    "        max_y = max(all_y_vals)\n",
    "        print(min_y)\n",
    "        print(max_y)\n",
    "        raise\n",
    "\n",
    "    if ASSEMBLER == \"verkko\":\n",
    "    \n",
    "        axes.set_xlim(51,58)\n",
    "        axes.set_ylim(48,63)\n",
    "    \n",
    "        axes.plot(\n",
    "            np.arange(48, 64),\n",
    "            np.arange(48, 64),\n",
    "            ls=\"dotted\",\n",
    "            lw=1,\n",
    "            zorder=0,\n",
    "            c=\"lightgrey\"\n",
    "        )\n",
    "\n",
    "    if ASSEMBLER == \"hifiasm\":\n",
    "        \n",
    "        axes.set_xlim(48,60)\n",
    "        axes.set_ylim(48,62)\n",
    "    \n",
    "        axes.plot(\n",
    "            np.arange(48, 64),\n",
    "            np.arange(48, 64),\n",
    "            ls=\"dotted\",\n",
    "            lw=1,\n",
    "            zorder=0,\n",
    "            c=\"lightgrey\"\n",
    "        )\n",
    "        \n",
    "    return axes\n",
    "\n",
    "\n",
    "def create_sample_qv_scatter(skip_plot=False):\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(8,8))\n",
    "    ax = plot_sample_vs_sample_qv(ax)\n",
    "    for ext in DEFAULT_PLOT_EXT:\n",
    "        if not skip_plot:\n",
    "            out_path = _MY_OUT_PATH.joinpath(f\"fig1_panel_qv-est.{ASSEMBLER}.{ext}\")\n",
    "            save_figure(out_path, fig)\n",
    "    plt.close()\n",
    "    return plot_sample_vs_sample_qv\n",
    "        \n",
    "cache_qv_estimate_table(aggregate=\"sequence\")\n",
    "cache_qv_estimate_table(aggregate=\"sample\")\n",
    "\n",
    "get_qv_panel = create_sample_qv_scatter(True)\n",
    "\n",
    "#_ = plot_unassigned_stats()\n",
    "#_ = plot_unassigned_vs_read_stats(\"ontul_cov\", \"ONT-UL cov. (x-fold)\")\n",
    "#_ = plot_unassigned_vs_read_stats(\"hifi_cov\", \"HiFi cov. (x-fold)\")\n",
    "#_ = plot_unassigned_vs_read_stats(\"ontul_n50\", \"ONT-UL N50 (kbp)\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
