


# Phasing

This contains the pipeline used to run the phasing experiments for the HGSVC3 paper.
See the steps below to replicate the analysis.

## How to replicate the results


### Step 1: Downloading input files

Reproduce PanGenie results using the pipeline https://github.com/eblerjana/hgsvc3/tree/main/experiments/genotyping or download PanGenie 1kg genotypes and panel for HGSVC3 data from:
```bat
wget -P inputs/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC3/release/Genotyping_1kGP/20240716_PanGenie-genotypes/pangenie_chm13_all_decomposed_lenient.vcf.gz
wget -P inputs/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC3/release/Genotyping_1kGP/20240716_PanGenie-genotypes/MC_hgsvc3-hprc_chm13_filtered_decomposed.vcf.gz
```


Download CHM13 reference genome:

```bat
wget -P inputs/ https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/analysis_set/chm13v2.0.fa.gz
gunzip inputs/chm13v2.0.fa.gz
```

Download GRCh38 reference genome:

```bat
wget -P inputs/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/technical/reference/20200513_hg38_NoALT/hg38.no_alt.fa.gz
gunzip inputs/hg38.no_alt.fa.gz
```


Download all genetic maps for CHM13 `` chr*.t2t.gmap.resorted.gmap.gz" `` from: [https://github.com/JosephLalli/phasing_T2T/tree/master/t2t_lifted_chrom_maps](https://github.com/JosephLalli/phasing_T2T/tree/master/resources/recombination_maps/t2t_lifted_from_hapmap) and put them to ``inputs/``


Download Illumina-based SNP and indel calls using the provided Snakefile:

```bat
cd inputs
snakemake
```

Download phased Illumina-based SNP and indel calls

```bat
wget -P inputs/ https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/variants/1000_Genomes_Project/chm13v2.0/Phased_SHAPEIT5_v1.1/1KGP.CHM13v2.0.whole_genome.recalibrated.snp_indel.pass.phased.native_maps.biallelic.3202.bcf.gz
``` 

Download NYGC phased panel (GRCh38) VCFs from http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20220422_3202_phased_SNV_INDEL_SV/ and combine them into a single VCF called `` inputs/1kGP_high_coverage_Illumina.filtered.SNV_INDEL_SV_phased_panel.vcf.gz `` using `` bcftools concat ``

Get missing BED files:
```bat
wget -P inputs/ https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v3.3/CHM13@all/Union/CHM13_notinalldifficultregions.bed.gz
gunzip inputs/CHM13_notinalldifficultregions.bed.gz

wget -P inputs/ https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v3.3/CHM13@all/LowComplexity/CHM13_AllTandemRepeatsandHomopolymers_slop5.bed.gz
gunzip inputs/CHM13_AllTandemRepeatsandHomopolymers_slop5.bed.gz
```


Furthermore, 1000G Illumina reads for all 3,202 samples and additional panel samples need to be provided (one file per sample). Paths to these files need to be provided in the `` inputs/1kg-reads-updated.tsv `` (this file needs to be edited).
Reads were obtained from the locations below and one FASTA file per sample was generated by concatenating individual FASTA/FASTQ files into a single file.

* **3,202 1kg samples:** http://ftp.sra.ebi.ac.uk/vol1/fastq/
* **NA24385:** https://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/NIST_Illumina_2x250bps/reads/
* **NA21309:** https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC_PLUS/NA21309/raw_data/Illumina/child/
* **HG01123, HG02486, HG02559:** https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=submissions/30E441F3-6820-4BF6-BCF4-E64D56C8D6A4--TRUSEQ/



### Step 2: run snakemake pipeline
Dependencies:
* snakemake
* singularity
* conda

```bat
snakemake -j <nr_cores> --use-conda
```
* The phased variants can be found in: ``results-with-rare/shapeit/shapeit/phased_shapeit.vcf.gz ``
* The consensus haplotypes of all 3,202 samples will be written to: `` results-with-rare/consensus-haplotypes/haplotypes/PanGenie-SHAPEIT/all/ ``

